{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gathering pathways, resolving gene homologs and preparing to visualize.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import copy\n",
    "import json\n",
    "import re\n",
    "import pickle as pkl\n",
    "from importlib import reload\n",
    "from zipfile import ZipFile\n",
    "from io import BytesIO\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from owlready2 import *\n",
    "import rdflib\n",
    "# from libsbml import *\n",
    "# import networkx as nx\n",
    "# from pyvis.network import Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31826\n"
     ]
    }
   ],
   "source": [
    "# pc = get_ontology(\"../data/PathwayCommons11.reactome.BIOPAX.owl\").load()\n",
    "# try:\n",
    "#   pc.destroy()\n",
    "# except:\n",
    "#   pass\n",
    "# pc = get_ontology(\"_all_\")\n",
    "# with open(\"../data/PathwayCommons12.inoh.BIOPAX.owl\", 'rb') as f:\n",
    "#   pc.load(fileobj=f)\n",
    "# with open(\"../data/PathwayCommons12.pathbank.BIOPAX.owl\", 'rb') as f:\n",
    "#   pc.load(fileobj=f)\n",
    "pc = get_ontology(\"../data/PathwayCommons12.reactome.BIOPAX.owl\").load()\n",
    "pc = get_ontology(\"../data/PathwayCommons12.kegg.BIOPAX.owl\").load()\n",
    "pc = get_ontology(\"../data/PathwayCommons12.humancyc.BIOPAX.owl\").load()\n",
    "pc = get_ontology(\"../data/PathwayCommons12.panther.BIOPAX.owl\").load()\n",
    "pc = get_ontology(\"../data/PathwayCommons12.inoh.BIOPAX.owl\").load()\n",
    "pc = get_ontology(\"../data/PathwayCommons12.pathbank.BIOPAX.owl\").load()\n",
    "pc = get_ontology(\"../data/PathwayCommons12.netpath.BIOPAX.owl\").load()\n",
    "pc = get_ontology(\"../data/PathwayCommons12.pid.BIOPAX.owl\").load()\n",
    "\n",
    "\n",
    "# reader = SBMLReader()\n",
    "# with ZipFile('../data/pathbank_primary_sbml.zip') as fzip:\n",
    "#     with fzip.open(fzip.namelist()[0]) as f:\n",
    "#       document = reader.readSBMLFromString(f.read().decode('utf-8'))\n",
    "\n",
    "# g = rdflib.Graph()\n",
    "# with ZipFile('../data/pathbank_primary_biopax.zip') as fzip:\n",
    "#   with fzip.open(fzip.namelist()[0]) as f:\n",
    "# #     fid = fzip.namelist()[0].split('.')[-2]\n",
    "    \n",
    "#     def sub_fn(matchobj):\n",
    "#       attr = matchobj.group(1)\n",
    "#       ref = matchobj.group(2)\n",
    "#       if ref.startswith('http'):\n",
    "#         # Skip\n",
    "#         return matchobj.group(0)\n",
    "#       ref = ref.replace('/', '_')\n",
    "#       if ref.startswith('#'):\n",
    "#         # Leave '#' at the front\n",
    "#         ref = f'#{ref[1:]}'\n",
    "#       else:\n",
    "#         ref = f'{ref}'\n",
    "#       return f'rdf:{attr}=\"{ref}\"'\n",
    "    \n",
    "#     fdata = f.read().decode('utf-8')\n",
    "#     fixed_fdata = re.sub('rdf:(ID) ?= ?\"(.*?)\"', sub_fn, fdata)\n",
    "#     fixed_fdata = re.sub('rdf:(resource) ?= ?\"(.*?)\"', sub_fn, fixed_fdata)\n",
    "# #     fixed_fdata = fixed_fdata.replace('xmlns=\"http://smpdb.ca/pathways/#\"', '')\n",
    "#     fdata_io = BytesIO(fixed_fdata.encode('utf-8'))\n",
    "# #     print(fixed_fdata)\n",
    "#     g.parse(data=fixed_fdata)\n",
    "# #     pc = get_ontology('../data/pathbank_primary_biopax.zip').load(fileobj=fdata_io)\n",
    "\n",
    "# i = 0\n",
    "# for root, dirs, files in os.walk(\"../data/pathbank_primary_biopax\"):\n",
    "#   for name in files:\n",
    "#     if not name.endswith('.owl'):\n",
    "#       continue\n",
    "#     fpath = os.path.join(root, name)\n",
    "#     print(fpath)\n",
    "#     pc = get_ontology(fpath).load()\n",
    "#     i += 1\n",
    "#     if i >= 1:\n",
    "#       break\n",
    "\n",
    "Control = list(filter(lambda x: x.name == 'Control', pc.world.classes()))[0]\n",
    "print(len(Control.instances()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load PANTHER gene orthologs\n",
    "```\n",
    "ftp://ftp.pantherdb.org/ortholog/14.1/RefGenomeOrthologs.tar.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>Type</th>\n",
       "      <th>Common ancestor</th>\n",
       "      <th>something else</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARATH|TAIR=AT4G37920|UniProtKB=Q84WN0</td>\n",
       "      <td>ARATH|TAIR=AT1G36320|UniProtKB=Q9C8X8</td>\n",
       "      <td>P</td>\n",
       "      <td>Embryophyta|Magnoliophyta</td>\n",
       "      <td>PTHR31755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HUMAN|HGNC=10663|UniProtKB=O60524</td>\n",
       "      <td>MOUSE|MGI=MGI=1918305|UniProtKB=Q8CCP0</td>\n",
       "      <td>LDO</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>PTHR15239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HUMAN|HGNC=10663|UniProtKB=O60524</td>\n",
       "      <td>RAT|Ensembl=ENSRNOG00000056128|UniProtKB=A0A0G...</td>\n",
       "      <td>LDO</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>PTHR15239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HUMAN|HGNC=10663|UniProtKB=O60524</td>\n",
       "      <td>CHICK|Ensembl=ENSGALG00000012263|UniProtKB=F1N8T0</td>\n",
       "      <td>LDO</td>\n",
       "      <td>Amniota</td>\n",
       "      <td>PTHR15239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HUMAN|HGNC=10663|UniProtKB=O60524</td>\n",
       "      <td>DANRE|Ensembl=ENSDARG00000102859|UniProtKB=A0A...</td>\n",
       "      <td>LDO</td>\n",
       "      <td>Euteleostomi</td>\n",
       "      <td>PTHR15239</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Gene  \\\n",
       "0  ARATH|TAIR=AT4G37920|UniProtKB=Q84WN0   \n",
       "1      HUMAN|HGNC=10663|UniProtKB=O60524   \n",
       "2      HUMAN|HGNC=10663|UniProtKB=O60524   \n",
       "3      HUMAN|HGNC=10663|UniProtKB=O60524   \n",
       "4      HUMAN|HGNC=10663|UniProtKB=O60524   \n",
       "\n",
       "                                            Ortholog Type  \\\n",
       "0              ARATH|TAIR=AT1G36320|UniProtKB=Q9C8X8    P   \n",
       "1             MOUSE|MGI=MGI=1918305|UniProtKB=Q8CCP0  LDO   \n",
       "2  RAT|Ensembl=ENSRNOG00000056128|UniProtKB=A0A0G...  LDO   \n",
       "3  CHICK|Ensembl=ENSGALG00000012263|UniProtKB=F1N8T0  LDO   \n",
       "4  DANRE|Ensembl=ENSDARG00000102859|UniProtKB=A0A...  LDO   \n",
       "\n",
       "             Common ancestor something else  \n",
       "0  Embryophyta|Magnoliophyta      PTHR31755  \n",
       "1           Euarchontoglires      PTHR15239  \n",
       "2           Euarchontoglires      PTHR15239  \n",
       "3                    Amniota      PTHR15239  \n",
       "4               Euteleostomi      PTHR15239  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orthologs = pd.read_csv(\n",
    "  '../data/RefGenomeOrthologs', delimiter='\\t', \n",
    "  names=['Gene', 'Ortholog', 'Type', 'Common ancestor', 'something else']\n",
    ")\n",
    "orthologs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build an ortholog -> mouse gene map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_db_name(db):\n",
    "  if db == 'UniProtKB' or db == 'uniprot knowledgebase':\n",
    "    return 'uniprot'\n",
    "  elif db == 'ncbi gene':\n",
    "    return 'ncbi'\n",
    "  elif db == 'hgnc symbol':\n",
    "    return 'hgnc'\n",
    "  return db.lower()\n",
    "\n",
    "\n",
    "def convert_db_id(id):\n",
    "  if ':' in str(id):\n",
    "    return id.split(':')[-1].lower()\n",
    "  return str(id).lower()\n",
    "\n",
    "\n",
    "def make_db_ref(db, id):\n",
    "  return '{0}:{1}'.format(convert_db_name(db), convert_db_id(id))\n",
    "\n",
    "\n",
    "try:\n",
    "  # Load precomputed ortholog gene map\n",
    "  with open('../data/gene-map-all2mouse', 'rb') as f:\n",
    "    all2mouse = pkl.load(f)\n",
    "except:\n",
    "  # Compute and store the ortholog gene map\n",
    "  all2mouse = {}\n",
    "  for _, row in orthologs.iterrows():\n",
    "    # Keep only the closest homologs\n",
    "    if row['Type'] != 'LDO':\n",
    "      continue\n",
    "    if not row['Ortholog'].startswith('MOUSE|'):\n",
    "      continue\n",
    "\n",
    "    dst = [\n",
    "      make_db_ref(d.split('=')[0], d.split('=')[-1])\n",
    "      for d in row['Ortholog'].split('|')[1:]\n",
    "    ]\n",
    "\n",
    "    for src in row['Gene'].split('|')[1:]:\n",
    "      src_parts = src.split('=')\n",
    "      db = src_parts[0]\n",
    "      id = src_parts[-1]\n",
    "      all2mouse[make_db_ref(db, id)] = dst\n",
    "\n",
    "  with open('../data/gene-map-all2mouse', 'wb') as f:\n",
    "    pkl.dump(all2mouse, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load mouse gene info from NCBI Gene\n",
    "```\n",
    "ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia/Mus_musculus.gene_info.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_gene_info(path):\n",
    "  \"\"\"Load the NCBI Gene file and returns a {gene_name: info} dictionary\"\"\"\n",
    "  gene_info = pd.read_csv(path, delimiter='\\t')\n",
    "  gene_info_map = {}\n",
    "  gene_name_map = {}\n",
    "  for _, row in gene_info.iterrows():\n",
    "    info = {\n",
    "      'name': row['Symbol'],\n",
    "#       'type': row['type_of_gene'],\n",
    "#       'desc': row['Other_designations'].split('|')[0],\n",
    "    }\n",
    "    ids = [make_db_ref('ncbi', row['GeneID'])]\n",
    "    gene_name_map[info['name']] = info\n",
    "    for xref in row['dbXrefs'].split('|'):\n",
    "      xref_parts = xref.split(':')\n",
    "      db = xref_parts[0]\n",
    "      id = xref_parts[-1]\n",
    "      ids.append(make_db_ref(db, id))\n",
    "    for id in ids:\n",
    "      gene_info_map[id] = info\n",
    "#     for s in row['Synonyms'].split('|'):\n",
    "#       gene_info_map[s] = info\n",
    "  return gene_info, gene_info_map, gene_name_map\n",
    "\n",
    "gene_info, gene_info_map, gene_name_map = load_gene_info('../data/Mus_musculus.gene_info')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exploring the BIOPAX format here.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # model = document.getModel()\n",
    "# # model.getNumSpecies()\n",
    "# # model.getNumReactions()\n",
    "# # r = model.reactions[0]\n",
    "# # dir(r)\n",
    "# # dir(model.compartments[0])\n",
    "# # r.reactants\n",
    "# dir(g)\n",
    "# for s, p, o in g:\n",
    "#   print(p, o)\n",
    "# #   break\n",
    "# # list(pc.world.classes())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pathways 5772\n",
      "interactions 70043\n",
      "['INDIRECT_get_properties', '__class__', '__classcell__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_equivalent_to', '_get_instance_possible_relations', '_get_is_instance_of', '_instance_equivalent_to_changed', '_instance_is_a_changed', '_name', '_set_is_instance_of', 'comment', 'dataSource', 'differents', 'displayName', 'generate_default_name', 'get_equivalent_to', 'get_inverse_properties', 'get_iri', 'get_name', 'get_properties', 'iri', 'is_a', 'is_instance_of', 'name', 'namespace', 'organism', 'pathwayComponent', 'pathwayOrder', 'set_equivalent_to', 'set_iri', 'set_name', 'storid', 'xref']\n",
      "Pathway\n",
      "{biopax-level3.displayName, biopax-level3.organism, biopax-level3.dataSource, biopax-level3.comment, biopax-level3.xref, biopax-level3.pathwayComponent, biopax-level3.pathwayOrder}\n",
      "Intestinal saccharidase deficiencies\n",
      "[.PathwayStep_dd2033278606d885216c78bed351ce77, .PathwayStep_e3e33599b25eb1a5fc96a98a5f2e0ec5, .PathwayStep_b6ebcc24cc2762b72ebae14c15955efe, .PathwayStep_8c8e181776d7005e041c9c72ed4b6778, .PathwayStep_a0d0fc06172b71ccf332e14cb86cf381]\n",
      "{biopax-level3.stepProcess, biopax-level3.comment}\n",
      "[reactome.R-HSA-5658001, .Catalysis_81c99aaa8b59ab2e799a448799c57ddc]\n",
      "None\n",
      "None\n",
      "[]\n",
      "reactome.R-HSA-5658001\n",
      "{biopax-level3.displayName, biopax-level3.conversionDirection, biopax-level3.dataSource, biopax-level3.comment, biopax-level3.xref, biopax-level3.left, biopax-level3.eCNumber}\n",
      "Defective LCT does not hydrolyze Lac\n",
      "\n",
      "{biopax-level3.displayName, biopax-level3.name, biopax-level3.entityReference, biopax-level3.cellularLocation, biopax-level3.dataSource, biopax-level3.comment, biopax-level3.xref}\n",
      "H2O\n",
      "chebi.CHEBI:15377\n",
      "{biopax-level3.id, biopax-level3.comment, biopax-level3.idVersion, biopax-level3.db}\n",
      "R-ALL-109276\n",
      "reactome\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "Pathway = list(filter(lambda x: x.name == 'Pathway', pc.world.classes()))[0]\n",
    "Interaction = list(filter(lambda x: x.name == 'Interaction', pc.world.classes()))[0]\n",
    "Conversion = list(filter(lambda x: x.name == 'Conversion', pc.world.classes()))[0]\n",
    "Control = list(filter(lambda x: x.name == 'Control', pc.world.classes()))[0]\n",
    "Catalysis = list(filter(lambda x: x.name == 'Catalysis', pc.world.classes()))[0]\n",
    "TemplateReaction = list(filter(lambda x: x.name == 'TemplateReaction', pc.world.classes()))[0]\n",
    "BiochemicalReaction = list(filter(lambda x: x.name == 'BiochemicalReaction', pc.world.classes()))[0]\n",
    "BindingFeature = list(filter(lambda x: x.name == 'BindingFeature', pc.world.classes()))[0]\n",
    "ModificationFeature = list(filter(lambda x: x.name == 'ModificationFeature', pc.world.classes()))[0]\n",
    "Complex = list(filter(lambda x: x.name == 'Complex', pc.world.classes()))[0]\n",
    "print(f'pathways {len(Pathway.instances())}')\n",
    "print(f'interactions {len(Interaction.instances())}')\n",
    "for p in Pathway.instances()[0:1]:\n",
    "  print(dir(p))\n",
    "  print(str(p.__class__).replace('biopax-level3.', ''))\n",
    "  print(p.get_properties())\n",
    "  print(p.displayName)\n",
    "  print(p.pathwayOrder)\n",
    "  print(p.pathwayOrder[0].get_properties())\n",
    "  print(p.pathwayOrder[0].stepProcess)\n",
    "  print(p.pathwayOrder[0].stepConversion)\n",
    "  print(p.pathwayOrder[0].stepDirection)\n",
    "  print(p.pathwayOrder[0].nextStep)\n",
    "  print(p.pathwayComponent[0])\n",
    "  print(p.pathwayComponent[0].get_properties())\n",
    "  print(p.pathwayComponent[0].displayName)\n",
    "  print()\n",
    "  print(p.pathwayComponent[0].left[0].get_properties())\n",
    "  print(p.pathwayComponent[0].left[0].displayName)\n",
    "  print(p.pathwayComponent[0].left[0].entityReference)\n",
    "  for c in p.pathwayComponent[0].left[0].component:\n",
    "    print('--c')\n",
    "    print(c)\n",
    "    print(c.displayName)\n",
    "    if hasattr(c, 'component'):\n",
    "      for c2 in c.component:\n",
    "        print('--c2')\n",
    "        print(c2.get_properties())\n",
    "        print(c2)\n",
    "        print(c2.displayName)\n",
    "        if hasattr(c2, 'component'):\n",
    "          for c3 in c2.component:\n",
    "            print('--c3')\n",
    "            print(c3.get_properties())\n",
    "            print(c3)\n",
    "            print(c3.displayName)\n",
    "  print(p.pathwayComponent[0].left[0].xref[0].get_properties())\n",
    "  for xref in p.pathwayComponent[0].left[0].xref:\n",
    "    print(xref.id)\n",
    "    print(xref.db)\n",
    "#   print(p.pathwayComponent[0].left[0].entityReference.name)\n",
    "#   print(p.pathwayComponent[0].left[0].entityReference.standardName)\n",
    "#   print(p.pathwayComponent[0].left[0].entityReference.displayName)\n",
    "#   print(p.pathwayComponent[0].left[0].entityReference.xref[0].get_properties())\n",
    "\n",
    "#   print(p.pathwayComponent[0].controller)\n",
    "#   print(p.pathwayComponent[0].controller[0].get_properties())\n",
    "#   print(p.pathwayComponent[0].controller[0].displayName)\n",
    "#   print(p.pathwayComponent[0].controller[0].entityReference)\n",
    "#   print(p.pathwayComponent[0].controller[0].entityReference.get_properties())\n",
    "#   print(p.pathwayComponent[0].controller[0].entityReference.organism.displayName)\n",
    "#   print(p.pathwayComponent[0].controller[0].entityReference.organism)\n",
    "#   print()\n",
    "#   print(p.pathwayComponent[0].controlled)\n",
    "#   print(p.pathwayComponent[0].controlled.get_properties())\n",
    "#   print(p.pathwayComponent[0].controlled.displayName)\n",
    "#   print(hasattr(p.pathwayComponent[0].controlled, 'displayName'))\n",
    "#   print(p.pathwayComponent[0].controlled.conversionDirection)\n",
    "  print()\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].get_properties())\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].displayName)\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].entityReference)\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].entityReference.name)\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].entityReference.standardName)\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].entityReference.displayName)\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].entityReference.xref[0].get_properties())\n",
    "  print()\n",
    "#   print(p.pathwayComponent)\n",
    "\n",
    "#   print(p.xref)\n",
    "#   print(p.xref[0].get_properties())\n",
    "#   print(p.xref[0].db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"hgnc:10663\": [\n",
      "    \"mgi:1918305\",\n",
      "    \"uniprot:q8ccp0\"\n",
      "  ],\n",
      "  \"uniprot:o60524\": [\n",
      "    \"mgi:1918305\",\n",
      "    \"uniprot:q8ccp0\"\n",
      "  ],\n",
      "  \"hgnc:20854\": [\n",
      "    \"mgi:1914066\",\n",
      "    \"uniprot:q9d305\"\n",
      "  ],\n",
      "  \"un\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(all2mouse, indent=2)[:200])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse BIOPAX into a more actionable form\n",
    "Extracting pathways and resolving chemical entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31826\n",
      "1830\n",
      "32836\n",
      "66492\n"
     ]
    }
   ],
   "source": [
    "def parse_molecule(m):\n",
    "  res = {\n",
    "    'type': 'molecule',\n",
    "    'name': m.displayName if m.displayName is not None else m.name,\n",
    "#     'name': getattr(m, 'displayName', m.name),\n",
    "#     'name': m.name,\n",
    "  }\n",
    "  if isinstance(m, Complex):\n",
    "    res['type'] = 'complex'\n",
    "    res['component'] = [parse_molecule(cc) for cc in m.component]\n",
    "  if hasattr(m, 'cellularLocation') and m.cellularLocation is not None:\n",
    "    if len(m.cellularLocation.term) > 0:\n",
    "      res['cellularLocation'] = m.cellularLocation.term[0]\n",
    "#   if hasattr(m, 'feature') and m.feature is not None:\n",
    "#     res['feature'] = []\n",
    "#     if len(m.feature) > 0:\n",
    "#       for f in m.feature:\n",
    "#         if not isinstance(f, ModificationFeature) and not isinstance(f, BindingFeature):\n",
    "#           continue\n",
    "#         feature = {\n",
    "#           'biopaxType': str(f.__class__).replace('biopax-level3.', ''),\n",
    "#         }\n",
    "#         if hasattr(f, 'modificationType') and f.modificationType is not None:\n",
    "#           feature['modificationType'] = f.modificationType.term[0]\n",
    "#         if hasattr(f, 'featureLocation') and f.featureLocation is not None:\n",
    "#           feature['featureLocation'] = [{\n",
    "#             'sequencePosition': l.sequencePosition,\n",
    "#             'positionStatus': l.positionStatus,\n",
    "#           } for l in f.featureLocation]\n",
    "# #         print(f.get_properties())\n",
    "# #         print(f.modificationType.term[0])\n",
    "# #         print(f.featureLocation[0].get_properties())\n",
    "# #         print(f.featureLocation[0].sequencePosition)\n",
    "# #         print(f.featureLocation[0].positionStatus)\n",
    "# # #         print(f.comment)\n",
    "# #         print(f.featureLocation)\n",
    "# #         print(f.modificationType.term[0])\n",
    "#         res['feature'].append(feature)\n",
    "  if hasattr(m, 'entityReference') and m.entityReference is not None:\n",
    "    e = m.entityReference\n",
    "    res['entityReference'] = {\n",
    "      'name': getattr(e, 'displayName', e.name),\n",
    "    }\n",
    "    if hasattr(e, 'xref') and e.xref is not None:\n",
    "      for x in e.xref:\n",
    "        if x.db is None:\n",
    "          continue\n",
    "        conv_id = '{0}:{1}'.format(convert_db_name(x.db), convert_db_id(x.id))\n",
    "        if conv_id in all2mouse:\n",
    "          for db_ref in all2mouse[conv_id]:\n",
    "            if db_ref in gene_info_map:\n",
    "              res['entityReference']['gene'] = gene_info_map[db_ref]\n",
    "        if 'gene' not in res['entityReference']:\n",
    "          if res['entityReference']['name'] in gene_name_map:\n",
    "            res['entityReference']['gene'] = gene_name_map[res['entityReference']['name']]\n",
    "          elif res['name'] in gene_name_map:\n",
    "            res['entityReference']['gene'] = gene_name_map[res['name']]\n",
    "        \n",
    "        # TODO: resolve small molecules\n",
    "        if ('uniprot' in x.db) \\\n",
    "          or ( \\\n",
    "            (x.db == 'ensembl' or x.db == 'ncbi gene') \\\n",
    "            and ( \\\n",
    "              'xref' not in res['entityReference'] \\\n",
    "              or res['entityReference']['xref']['db'] != 'ncbi gene' \\\n",
    "            ) \\\n",
    "          ) \\\n",
    "          or ( \\\n",
    "#             x.db in ['chebi'] \\\n",
    "#             and \n",
    "            'xref' not in res['entityReference'] \\\n",
    "          ):\n",
    "          # NCBI Gene has the top priority and the Ensemlb goes next\n",
    "          res['entityReference']['xref'] = {\n",
    "            'id': convert_db_id(x.id),\n",
    "            'db': convert_db_name(x.db),\n",
    "          }\n",
    "#       res['entityReference']['xref'] = [\n",
    "#         {\n",
    "#           'id': x.id,\n",
    "#           'db': x.db,\n",
    "# #           'relationshipType': [\n",
    "# #              r.term\n",
    "# #             for r in getattr(x, 'relationshipType', [])\n",
    "# #           ],\n",
    "#         }\n",
    "#         for x in e.xref\n",
    "#       ]\n",
    "#     if hasattr(e, 'organism') and e.organism is not None:\n",
    "#       o = e.organism\n",
    "#       res['entityReference']['organism'] = {\n",
    "#         'name': getattr(o, 'displayName', o.name),\n",
    "#       }\n",
    "#       if 'Homo' not in res['entityReference']['organism']['name']:\n",
    "#         print(res['entityReference']['organism']['name'])\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_reaction(r):\n",
    "  res = {\n",
    "    'type': 'reaction',\n",
    "    'name': r.displayName if r.displayName is not None else r.name,\n",
    "#     'name': getattr(r, 'displayName', r.name),\n",
    "#     'comment': getattr(r, 'comment', None),\n",
    "    'conversionDirection': getattr(r, 'conversionDirection', None),\n",
    "  }\n",
    "  if hasattr(r, 'left') and r.left is not None:\n",
    "    res['left'] = [parse_molecule(c) for _, c in enumerate(r.left)]\n",
    "  if hasattr(r, 'right') and r.right is not None:\n",
    "    res['right'] = [parse_molecule(c) for _, c in enumerate(r.right)]\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_control(c):\n",
    "  res = {\n",
    "    'type': 'control',\n",
    "    'name': c.displayName if c.displayName is not None else c.name,\n",
    "#     'name': getattr(c, 'displayName', c.name),\n",
    "#     'comment': getattr(c, 'comment', None),\n",
    "    'controlType': getattr(c, 'controlType', None),\n",
    "  }\n",
    "  if hasattr(c, 'controller') and c.controller is not None:\n",
    "    res['controller'] = [parse_molecule(cc) for cc in c.controller]\n",
    "  if hasattr(c, 'cofactor') and c.cofactor is not None:\n",
    "    res['cofactor'] = [parse_molecule(cc) for cc in c.cofactor]\n",
    "  if hasattr(c, 'controlled') and c.controlled is not None:\n",
    "    res['controlled'] = parse_reaction(c.controlled)\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_template_reaction(c):\n",
    "  res = {\n",
    "    'type': 'template_reaction',\n",
    "    'name': c.displayName if c.displayName is not None else c.name,\n",
    "#     'name': getattr(c, 'displayName', c.name),\n",
    "#     'comment': getattr(c, 'comment', None),\n",
    "    'templateDirection': getattr(c, 'templateDirection', None),\n",
    "  }\n",
    "  if hasattr(c, 'template') and c.template is not None:\n",
    "    res['template'] = parse_molecule(c.template)\n",
    "  if hasattr(c, 'product') and c.product is not None:\n",
    "    res['product'] = [parse_molecule(cc) for cc in c.product]\n",
    "  return res\n",
    "\n",
    "\n",
    "parsed_pathway_steps = {}\n",
    "def parse_pathway_step(s):\n",
    "  reactions = []\n",
    "  if s.name in parsed_pathway_steps:\n",
    "    # Avoid infinite recursions\n",
    "    return reactions\n",
    "  parsed_pathway_steps[s.name] = ''\n",
    "  res = {\n",
    "    'type': 'pathway_step',\n",
    "    'name': s.displayName if s.displayName is not None else s.name,\n",
    "#     'name': getattr(c, 'displayName', c.name),\n",
    "#     'comment': getattr(c, 'comment', None),\n",
    "    'stepDirection': getattr(s, 'stepDirection', None),\n",
    "  }\n",
    "  if hasattr(s, 'stepConversion') and s.stepConversion is not None:\n",
    "    res['stepConversion'] = parse_reaction(s.stepConversion)\n",
    "    reactions.append(res['stepConversion'])\n",
    "  if hasattr(s, 'stepProcess') and s.stepProcess is not None:\n",
    "    res['stepProcess'] = [parse_control(cc) for cc in s.stepProcess]\n",
    "    reactions.extend(res['stepProcess'])\n",
    "  if hasattr(s, 'nextStep') and s.nextStep is not None:\n",
    "    for ss in s.nextStep:\n",
    "      reactions.extend(parse_pathway_step(ss))\n",
    "  return reactions\n",
    "\n",
    "def parse_pathway(p):\n",
    "  res = {\n",
    "    'type': 'pathway',\n",
    "    'name': getattr(p, 'displayName', p.name),\n",
    "#     'comment': getattr(p, 'comment', None),\n",
    "  }\n",
    "  if hasattr(p, 'pathwayComponent') and p.pathwayComponent is not None:\n",
    "    res['pathwayComponent'] = []\n",
    "    for c in p.pathwayComponent:\n",
    "      if isinstance(c, Pathway):\n",
    "#         res_list.append(parse_pathway(c))\n",
    "        continue\n",
    "      elif isinstance(c, Control):\n",
    "        res['pathwayComponent'].append(parse_control(c))\n",
    "      elif isinstance(c, TemplateReaction):\n",
    "        res['pathwayComponent'].append(parse_template_reaction(c))\n",
    "      else:\n",
    "        res['pathwayComponent'].append(parse_reaction(c))\n",
    "  \n",
    "  if hasattr(p, 'pathwayOrder') and p.pathwayOrder is not None:\n",
    "    if 'pathwayComponent' not in res:\n",
    "      res['pathwayComponent'] = []\n",
    "    for s in p.pathwayOrder:\n",
    "      res['pathwayComponent'].extend(parse_pathway_step(s))\n",
    "\n",
    "  return res\n",
    "\n",
    "\n",
    "# print(list(pc.world.properties()))\n",
    "\n",
    "parsed_map = {}\n",
    "pathways = []\n",
    "\n",
    "generic_pathway_template = {\n",
    "  'type': 'pathway',\n",
    "  'name': '<generic>',\n",
    "  'pathwayComponent': [],\n",
    "}\n",
    "\n",
    "def process_reactions(reaction_class, parse_fn):\n",
    "  print(len(reaction_class.instances()))\n",
    "  for c in reaction_class.instances()[:]:\n",
    "    if c.name in parsed_map:\n",
    "      continue\n",
    "    parsed_map[c.name] = ''\n",
    "#     for xref in c.xref:\n",
    "#       print(xref.title)\n",
    "#       print(xref.get_properties())\n",
    "#     print(dir(c))\n",
    "    r = parse_fn(c)\n",
    "#     print(r)\n",
    "  #   print(json.dumps(r, indent=4))\n",
    "    generic_pathway = copy.deepcopy(generic_pathway_template)\n",
    "    generic_pathway['pathwayComponent'].append(r)\n",
    "    pathways.append(generic_pathway) \n",
    "\n",
    "for reaction_class, parse_fn in [\n",
    "  (Control, parse_control),\n",
    "  (TemplateReaction, parse_template_reaction),\n",
    "#   (BiochemicalReaction, parse_reaction),\n",
    "  (Conversion, parse_reaction),\n",
    "#   (Interaction, parse_reaction),\n",
    "]:\n",
    "  process_reactions(reaction_class, parse_fn)\n",
    "\n",
    "# for p in Pathway.instances()[:]:\n",
    "#   pathway = parse_pathway(p)\n",
    "# #   print(json.dumps(res, indent=2)[:100000])\n",
    "#   pathways.append(pathway)\n",
    "\n",
    "print(len(pathways))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41886\n"
     ]
    }
   ],
   "source": [
    "# Filter out all the reactions without resolved genes\n",
    "filtered_pathways = []\n",
    "\n",
    "def gene_resolved(m):\n",
    "  if m['type'] == 'complex':\n",
    "    return any([gene_resolved(c) for c in m['component']])\n",
    "  if 'entityReference' not in m:\n",
    "    return False\n",
    "  return 'gene' in m['entityReference']\n",
    "\n",
    "for p in pathways:\n",
    "  filtered_p = copy.deepcopy(p)\n",
    "  filtered_p['pathwayComponent'] = []\n",
    "  for r in p['pathwayComponent']:\n",
    "    resolved = False\n",
    "    rr = r\n",
    "    if r['type'] == 'control':\n",
    "      for m in r['controller']:\n",
    "        resolved = resolved or gene_resolved(m)\n",
    "    elif r['type'] == 'template_reaction':\n",
    "      for m in r['product']:\n",
    "        resolved = resolved or gene_resolved(m)\n",
    "    if 'left' in rr:\n",
    "      for m in rr['left']:\n",
    "        resolved = resolved or gene_resolved(m)\n",
    "    if 'right' in rr:\n",
    "      for m in rr['right']:\n",
    "        resolved = resolved or gene_resolved(m)\n",
    "    if resolved:\n",
    "      filtered_p['pathwayComponent'].append(r)\n",
    "  if len(filtered_p['pathwayComponent']) > 0:\n",
    "    filtered_pathways.append(filtered_p)\n",
    "\n",
    "print(len(filtered_pathways))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert the parsed data into a graph form"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6\n",
      "7\n",
      "{\n",
      "    \"nodes\": [\n",
      "        {\n",
      "            \"__id\": 0,\n",
      "            \"type\": \"reaction\",\n",
      "            \"name\": \"Moderate kinase activity BRAF mutants:RAS:GTP homo/heterodimerize\"\n",
      "        },\n",
      "        {\n",
      "            \"__id\": 1,\n",
      "            \"type\": \"molecule\",\n",
      "            \"name\": \"BRAP_HUMAN\",\n",
      "            \"cellularLocation\": \"cytosol\",\n",
      "            \"entityReference\": {\n",
      "                \"name\": \"BRAP_HUMAN\",\n",
      "                \"gene\": {\n",
      "                    \"name\": \"Brap\"\n",
      "                },\n",
      "                \"xref\": {\n",
      "                    \"id\": \"q7z569\",\n",
      "                    \"db\": \"uniprot\"\n",
      "                }\n",
      "            }\n",
      "        },\n",
      "        {\n",
      "            \"__id\": 2,\n",
      "            \"type\": \"complex\",\n",
      "            \"name\": \"KSR1:MARK3\"\n",
      "        },\n",
      "        {\n",
      "            \"__id\": 3,\n",
      "            \"type\": \"complex\",\n",
      "            \"name\": \"RAS:GTP\"\n",
      "        },\n",
      "        {\n",
      "            \"__id\": 4,\n",
      "            \"type\": \"molecule\",\n",
      "            \"name\": \"phosphorylated moderate kinase activity BRAF mutants\",\n",
      "            \"cellularLocation\": \"cytosol\",\n",
      "            \"entityReference\": {\n",
      "                \"name\": \"phosphorylated moderate kinase activity BRAF mutants\",\n",
      "                \"xref\": {\n",
      "                    \"id\": \"r-hsa-8936658\",\n",
      "                    \"db\": \"reactome\"\n",
      "                }\n",
      "            }\n",
      "        },\n",
      "        {\n",
      "            \"__id\": 5,\n",
      "            \"type\": \"complex\",\n",
      "            \"name\": \"'receiver' RAF\"\n",
      "        }\n",
      "    ],\n",
      "    \"edges\": [\n",
      "        [\n",
      "            1,\n",
      "            0,\n",
      "            \"inhibitor\"\n",
      "        ],\n",
      "        [\n",
      "            2,\n",
      "            0,\n",
      "            \"inhibitor\"\n",
      "        ],\n",
      "        [\n",
      "            3,\n",
      "            0,\n",
      "            \"reactant\"\n",
      "        ],\n",
      "        [\n",
      "            4,\n",
      "            0,\n",
      "            \"reactant\"\n",
      "        ],\n",
      "        [\n",
      "            0,\n",
      "            3,\n",
      "            \"product\"\n",
      "        ],\n",
      "        [\n",
      "            0,\n",
      "            4,\n",
      "            \"product\"\n",
      "        ],\n",
      "        [\n",
      "            0,\n",
      "            5,\n",
      "            \"product\"\n",
      "        ]\n",
      "    ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "def location_group(data):\n",
    "  if 'cellularLocation' not in data:\n",
    "    return None\n",
    "  loc = data['cellularLocation']\n",
    "  if any([ l in loc for l in [\n",
    "    'extracellular region',\n",
    "    'external side of plasma membrane',\n",
    "    'cell junction',\n",
    "  ] ]):\n",
    "    return 5\n",
    "  elif any([ l in loc for l in [\n",
    "    'cell outer membrane',\n",
    "    'cell wall',\n",
    "    'cytoplasmic side of plasma membrane',\n",
    "    'plasma membrane',\n",
    "  ] ]):\n",
    "    return 4\n",
    "  elif any([ l in loc for l in [\n",
    "    'mitochondrial',\n",
    "  ] ]):\n",
    "    return 3\n",
    "  elif any([ l in loc for l in [\n",
    "    'cytosol',\n",
    "    'host cell cytosol',\n",
    "  ] ]):\n",
    "    return 0\n",
    "  elif any([ l in loc for l in [\n",
    "    'Golgi',\n",
    "  ] ]):\n",
    "    return -3\n",
    "  elif any([ l in loc for l in [\n",
    "    'endoplasmic reticulum',\n",
    "  ] ]):\n",
    "    return -4\n",
    "  elif any([ l in loc for l in [\n",
    "    'nuclear envelope',\n",
    "    'nucleoplasm',\n",
    "    'chromosome',\n",
    "    'chromosome, centromeric region',\n",
    "  ] ]):\n",
    "    return -5\n",
    "  return None\n",
    "\n",
    "def pathways2graph(pathways):\n",
    "  nodes = []\n",
    "  node_name_map = {}\n",
    "  edges = []\n",
    "  edge_map = {}\n",
    "  \n",
    "  def add_node(data):\n",
    "    name = data['entityReference']['name'] if 'entityReference' in data else data['name']\n",
    "    map_id = f'{name}@{location_group(data)}'\n",
    "    if map_id in node_name_map:\n",
    "      return node_name_map[map_id]\n",
    "    id = len(nodes)\n",
    "    node = {\n",
    "      '__id': id,\n",
    "      'type': data['type'],\n",
    "      'name': name,\n",
    "    }\n",
    "    if data['type'] == 'molecule':\n",
    "      if 'cellularLocation' in data:\n",
    "        node['cellularLocation'] = data['cellularLocation'].lower()\n",
    "      if 'entityReference' in data:\n",
    "        node['entityReference'] = data['entityReference']\n",
    "    nodes.append(node)\n",
    "    node_name_map[map_id] = id\n",
    "    return id\n",
    "  \n",
    "  def add_nodes(data):\n",
    "    ids = []\n",
    "    if data['type'] == 'complex':\n",
    "      ids.extend([ add_node(c) for c in data['component'] ])\n",
    "    else:\n",
    "      ids.append(add_node(data))\n",
    "    return list(set(ids))\n",
    "  \n",
    "  def add_edge(source, target, relation):\n",
    "    edge_key = f'{source}|{target}'\n",
    "    if edge_key in edge_map:\n",
    "      return\n",
    "    edges.append([source, target, relation])\n",
    "    edge_map[edge_key] = ''\n",
    "  \n",
    "  control_type2relation = {\n",
    "    'ACTIVATION': 'activator',\n",
    "    'INHIBITION': 'inhibitor',\n",
    "  }\n",
    "  \n",
    "  for pathway in pathways:\n",
    "    for reaction_container in pathway['pathwayComponent']:\n",
    "      reaction = reaction_container\n",
    "      \n",
    "      if reaction_container['type'] == 'control':\n",
    "        if 'controlled' not in reaction_container:\n",
    "          continue\n",
    "        reaction = reaction_container['controlled']\n",
    "    \n",
    "      reaction_node_id = add_node(reaction)\n",
    "      if reaction_container['type'] == 'control':\n",
    "        for c in reaction_container['controller']:\n",
    "          molecule_node_ids = add_nodes(c)\n",
    "          for m in molecule_node_ids:\n",
    "            relation = 'controller'\n",
    "            if reaction_container['controlType'] is not None:\n",
    "              relation = reaction_container['controlType']\n",
    "              for c2r in control_type2relation:\n",
    "                if c2r in relation:\n",
    "                  relation = control_type2relation[c2r]\n",
    "            add_edge(m, reaction_node_id, relation)\n",
    "      elif reaction_container['type'] == 'template_reaction':\n",
    "        if 'template' in reaction_container:\n",
    "          molecule_node_ids = add_nodes(reaction_container['template'])\n",
    "          [ add_edge(m, reaction_node_id, 'template') for m in molecule_node_ids ]\n",
    "        for product in reaction_container['product']:\n",
    "          molecule_node_ids = add_nodes(product)\n",
    "          [ add_edge(reaction_node_id, m, 'product') for m in molecule_node_ids ]\n",
    "          \n",
    "      if 'left' in reaction:\n",
    "        for molecule in reaction['left']:\n",
    "          molecule_node_ids = add_nodes(molecule)\n",
    "          if reaction['conversionDirection'] == 'LEFT_TO_RIGHT':\n",
    "            [ add_edge(m, reaction_node_id, 'reactant') for m in molecule_node_ids ]\n",
    "          elif reaction['conversionDirection'] == 'RIGHT_TO_LEFT':\n",
    "            [ add_edge(reaction_node_id, m, 'product') for m in molecule_node_ids ]\n",
    "          else:\n",
    "            [ add_edge(m, reaction_node_id, 'reactant') for m in molecule_node_ids ]\n",
    "          \n",
    "      if 'right' in reaction:\n",
    "        for molecule in reaction['right']:\n",
    "          molecule_node_ids = add_nodes(molecule)\n",
    "          if reaction['conversionDirection'] == 'LEFT_TO_RIGHT':\n",
    "            [ add_edge(reaction_node_id, m, 'product') for m in molecule_node_ids ]\n",
    "          elif reaction['conversionDirection'] == 'RIGHT_TO_LEFT':\n",
    "            [ add_edge(m, reaction_node_id, 'reactant') for m in molecule_node_ids ]\n",
    "          else:\n",
    "            [ add_edge(reaction_node_id, m, 'product') for m in molecule_node_ids ]\n",
    "        \n",
    "  return {\n",
    "    'nodes': nodes,\n",
    "    'edges': edges,\n",
    "  }\n",
    "\n",
    "print(len(pathways2graph(filtered_pathways[:1])['nodes']))\n",
    "print(len(pathways2graph(filtered_pathways[:1])['edges']))\n",
    "print(json.dumps(pathways2graph(filtered_pathways[0:1]), indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41886\n"
     ]
    }
   ],
   "source": [
    "print(len(filtered_pathways))\n",
    "with open('./results/pathways.json', 'w') as f:\n",
    "  json.dump(pathways2graph(filtered_pathways[:]), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"type\": \"pathway\",\n",
      "    \"name\": \"<generic>\",\n",
      "    \"pathwayComponent\": [\n",
      "        {\n",
      "            \"type\": \"control\",\n",
      "            \"name\": \"R-HSA-8936737\",\n",
      "            \"controlType\": \"INHIBITION\",\n",
      "            \"controller\": [\n",
      "                {\n",
      "                    \"type\": \"complex\",\n",
      "                    \"name\": \"BRAP:KSR1:MARK3\",\n",
      "                    \"component\": [\n",
      "                        {\n",
      "                            \"type\": \"molecule\",\n",
      "                            \"name\": \"BRAP\",\n",
      "                            \"cellularLocation\": \"cytosol\",\n",
      "                            \"entityReference\": {\n",
      "                                \"name\": \"BRAP_HUMAN\",\n",
      "                                \"gene\": {\n",
      "                                    \"name\": \"Brap\"\n",
      "                                },\n",
      "                                \"xref\": {\n",
      "                                    \"id\": \"q7z569\",\n",
      "                                    \"db\": \"uniprot\"\n",
      "                                }\n",
      "                            }\n",
      "                        },\n",
      "                        {\n",
      "                            \"type\": \"complex\",\n",
      "                            \"name\": \"KSR1:MARK3\",\n",
      "                            \"component\": [\n",
      "                                {\n",
      "                                    \"type\": \"molecule\",\n",
      "                                    \"name\": \"EMK2\",\n",
      "                                    \"cellularLocation\": \"cytosol\",\n",
      "                                    \"entityReference\": {\n",
      "                                        \"name\": \"MARK3_HUMAN\",\n",
      "                                        \"gene\": {\n",
      "                                            \"name\": \"Mark3\"\n",
      "                                        },\n",
      "                                        \"xref\": {\n",
      "                                            \"id\": \"p27448\",\n",
      "                                            \"db\": \"uniprot\"\n",
      "                                        }\n",
      "                                    }\n",
      "                                },\n",
      "                                {\n",
      "                                    \"type\": \"molecule\",\n",
      "                                    \"name\": \"KSR1\",\n",
      "                                    \"cellularLocation\": \"cytosol\",\n",
      "                                    \"entityReference\": {\n",
      "                                        \"name\": \"KSR1_HUMAN\",\n",
      "                                        \"xref\": {\n",
      "                                            \"id\": \"q8ivt5\",\n",
      "                                            \"db\": \"uniprot\"\n",
      "                                        },\n",
      "                                        \"gene\": {\n",
      "                                            \"name\": \"Ksr1\"\n",
      "                                        }\n",
      "                                    }\n",
      "                                }\n",
      "                            ],\n",
      "                            \"cellularLocation\": \"cytosol\"\n",
      "                        }\n",
      "                    ],\n",
      "                    \"cellularLocation\": \"cytosol\"\n",
      "                }\n",
      "            ],\n",
      "            \"cofactor\": [],\n",
      "            \"controlled\": {\n",
      "                \"type\": \"reaction\",\n",
      "                \"name\": \"Moderate kinase activity BRAF mutants:RAS:GTP homo/heterodimerize\",\n",
      "                \"conversionDirection\": \"LEFT_TO_RIGHT\",\n",
      "                \"left\": [\n",
      "                    {\n",
      "                        \"type\": \"complex\",\n",
      "                        \"name\": \"'receiver' RAF\",\n",
      "                        \"component\": [],\n",
      "                        \"cellularLocation\": \"cytosol\"\n",
      "                    },\n",
      "                    {\n",
      "                        \"type\": \"complex\",\n",
      "                        \"name\": \"p21 RAS:GTP:moderate kinase activity BRAF mutants\",\n",
      "                        \"component\": [\n",
      "                            {\n",
      "                                \"type\": \"complex\",\n",
      "                                \"name\": \"RAS:GTP\",\n",
      "                                \"component\": [\n",
      "                                    {\n",
      "                                        \"type\": \"molecule\",\n",
      "                                        \"name\": \"RAS\",\n",
      "                                        \"cellularLocation\": \"plasma membrane\",\n",
      "                                        \"entityReference\": {\n",
      "                                            \"name\": \"RAS mutants\",\n",
      "                                            \"xref\": {\n",
      "                                                \"id\": \"r-hsa-6802608\",\n",
      "                                                \"db\": \"reactome\"\n",
      "                                            }\n",
      "                                        }\n",
      "                                    },\n",
      "                                    {\n",
      "                                        \"type\": \"molecule\",\n",
      "                                        \"name\": \"GTP\",\n",
      "                                        \"cellularLocation\": \"cytosol\",\n",
      "                                        \"entityReference\": {\n",
      "                                            \"name\": \"GTP\",\n",
      "                                            \"xref\": {\n",
      "                                                \"id\": \"37045\",\n",
      "                                                \"db\": \"chebi\"\n",
      "                                            }\n",
      "                                        }\n",
      "                                    }\n",
      "                                ],\n",
      "                                \"cellularLocation\": \"plasma membrane\"\n",
      "                            },\n",
      "                            {\n",
      "                                \"type\": \"molecule\",\n",
      "                                \"name\": \"moderate kinase activity BRAF mutants\",\n",
      "                                \"cellularLocation\": \"cytosol\",\n",
      "                                \"entityReference\": {\n",
      "                                    \"name\": \"phosphorylated moderate kinase activity BRAF mutants\",\n",
      "                                    \"xref\": {\n",
      "                                        \"id\": \"r-hsa-8936658\",\n",
      "                                        \"db\": \"reactome\"\n",
      "                                    }\n",
      "                                }\n",
      "                            }\n",
      "                        ],\n",
      "                        \"cellularLocation\": \"plasma membrane\"\n",
      "                    }\n",
      "                ],\n",
      "                \"right\": [\n",
      "                    {\n",
      "                        \"type\": \"complex\",\n",
      "                        \"name\": \"RAS:GTP:moderate kinase activity BRAF mutants:RAF:YWHAB\",\n",
      "                        \"component\": [\n",
      "                            {\n",
      "                                \"type\": \"complex\",\n",
      "                                \"name\": \"'receiver' RAF\",\n",
      "                                \"component\": [],\n",
      "                                \"cellularLocation\": \"cytosol\"\n",
      "                            },\n",
      "                            {\n",
      "                                \"type\": \"complex\",\n",
      "                                \"name\": \"RAS:GTP\",\n",
      "                                \"component\": [\n",
      "                                    {\n",
      "                                        \"type\": \"molecule\",\n",
      "                                        \"name\": \"RAS\",\n",
      "                                        \"cellularLocation\": \"plasma membrane\",\n",
      "                                        \"entityReference\": {\n",
      "                                            \"name\": \"RAS mutants\",\n",
      "                                            \"xref\": {\n",
      "                                                \"id\": \"r-hsa-6802608\",\n",
      "                                                \"db\": \"reactome\"\n",
      "                                            }\n",
      "                                        }\n",
      "                                    },\n",
      "                                    {\n",
      "                                        \"type\": \"molecule\",\n",
      "                                        \"name\": \"GTP\",\n",
      "                                        \"cellularLocation\": \"cytosol\",\n",
      "                                        \"entityReference\": {\n",
      "                                            \"name\": \"GTP\",\n",
      "                                            \"xref\": {\n",
      "                                                \"id\": \"37045\",\n",
      "                                                \"db\": \"chebi\"\n",
      "                                            }\n",
      "                                        }\n",
      "                                    }\n",
      "                                ],\n",
      "                                \"cellularLocation\": \"plasma membrane\"\n",
      "                            },\n",
      "                            {\n",
      "                                \"type\": \"molecule\",\n",
      "                                \"name\": \"moderate kinase activity BRAF mutants\",\n",
      "                                \"cellularLocation\": \"cytosol\",\n",
      "                                \"entityReference\": {\n",
      "                                    \"name\": \"phosphorylated moderate kinase activity BRAF mutants\",\n",
      "                                    \"xref\": {\n",
      "                                        \"id\": \"r-hsa-8936658\",\n",
      "                                        \"db\": \"reactome\"\n",
      "                                    }\n",
      "                                }\n",
      "                            }\n",
      "                        ],\n",
      "                        \"cellularLocation\": \"plasma membrane\"\n",
      "                    }\n",
      "                ]\n",
      "            }\n",
      "        }\n",
      "    ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# print(len(pathways2graph(filtered_pathways[:])['nodes']))\n",
    "# print(len(pathways2graph(filtered_pathways[:])['edges']))\n",
    "print(json.dumps(filtered_pathways[0], indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Basolateralmembrane' 'CAAX-protein geranylgeranyltransferase complex'\n",
      " 'CELL' 'COPII-coated ER to Golgi transport vesicle' 'Cell surface'\n",
      " 'ER to Golgi transport vesicle membrane' 'Endocytic Vesicle'\n",
      " 'Extracellular Space' 'GO:0031983' 'Golgi apparatus' 'Golgi lumen'\n",
      " 'Golgi membrane' 'Golgi to ER transport vesicle membrane'\n",
      " 'Golgi trans face' 'Golgi-associated vesicle'\n",
      " 'Golgi-associated vesicle lumen' 'Golgi-associated vesicle membrane'\n",
      " 'Intracellular vesicle' 'Late endosome' 'Mitochondrial membrane'\n",
      " 'Nuclues' 'Proteasome' 'actin cytoskeleton' 'adherens junction'\n",
      " 'apicolateral plasma membrane' 'autophagosome' 'autophagosome membrane'\n",
      " 'axon terminus' 'axonemal microtubule' 'azurophil granule'\n",
      " 'azurophil granule lumen' 'azurophil granule membrane'\n",
      " 'basement membrane' 'basolateral plasma membrane'\n",
      " 'beta-galactosidase complex' 'brush border' 'caveola' 'cell'\n",
      " 'cell cortex' 'cell junction' 'cell leading edge' 'cell outer membrane'\n",
      " 'cell wall' 'cell-cell junction' 'centrosome' 'chromatin' 'chromosome'\n",
      " 'chromosome, centromeric region' 'ciliary basal body' 'ciliary base'\n",
      " 'ciliary membrane' 'ciliary tip' 'cilium' 'clathrin coat of coated pit'\n",
      " 'clathrin coat of endocytic vesicle' 'clathrin-coated endocytic vesicle'\n",
      " 'clathrin-coated endocytic vesicle membrane' 'clathrin-coated vesicle'\n",
      " 'clathrin-sculpted acetylcholine transport vesicle lumen'\n",
      " 'clathrin-sculpted acetylcholine transport vesicle membrane'\n",
      " 'clathrin-sculpted gamma-aminobutyric acid transport vesicle lumen'\n",
      " 'clathrin-sculpted gamma-aminobutyric acid transport vesicle membrane'\n",
      " 'clathrin-sculpted monoamine transport vesicle lumen'\n",
      " 'clathrin-sculpted monoamine transport vesicle membrane'\n",
      " 'cleavage furrow' 'coated pit' 'cornified envelope' 'cytoplasm'\n",
      " 'cytoplasmic side of plasma membrane' 'cytoplasmic vesicle'\n",
      " 'cytoplasmic vesicle lumen' 'cytoplasmic vesicle membrane' 'cytoskeletal'\n",
      " 'cytoskeleton' 'cytosol' 'dendritic spine' 'dense core granule'\n",
      " 'early endosome' 'early endosome lumen' 'early endosome membrane'\n",
      " 'endocytic vesicle lumen' 'endocytic vesicle membrane'\n",
      " 'endolysosome lumen' 'endolysosome membrane' 'endoplasmic reticulum'\n",
      " 'endoplasmic reticulum ( extrinsic to internal side of membrane )'\n",
      " 'endoplasmic reticulum ( inner space )'\n",
      " 'endoplasmic reticulum ( integral to membrane )'\n",
      " 'endoplasmic reticulum lumen' 'endoplasmic reticulum membrane'\n",
      " 'endoplasmic reticulum quality control compartment'\n",
      " 'endoplasmic reticulum(membrane)'\n",
      " 'endoplasmic reticulum-Golgi intermediate compartment'\n",
      " 'endoplasmic reticulum-Golgi intermediate compartment membrane'\n",
      " 'endosome' 'endosome lumen' 'endosome membrane' 'exocytic vesicle'\n",
      " 'external side of plasma membrane' 'extracellular'\n",
      " 'extracellular exosome' 'extracellular matrix' 'extracellular region'\n",
      " 'extrachromosomal circular DNA'\n",
      " 'extrinsic component of cytoplasmic side of plasma membrane'\n",
      " 'extrinsic component of endoplasmic reticulum membrane'\n",
      " 'extrinsic component of external side of plasma membrane'\n",
      " 'ficolin-1-rich granule lumen' 'ficolin-1-rich granule membrane'\n",
      " 'filamentous actin' 'focal adhesion' 'growth cone' 'hemidesmosome'\n",
      " 'heterochromatin' 'host cell cytosol'\n",
      " 'inner kinetochore of condensed chromosome' 'inner membrane complex'\n",
      " 'insulin-responsive compartment'\n",
      " 'integral component of cytoplasmic side of endoplasmic reticulum membrane'\n",
      " 'integral component of endoplasmic reticulum membrane'\n",
      " 'integral component of lumenal side of endoplasmic reticulum membrane'\n",
      " 'integral component of plasma membrane' 'integral to membrane'\n",
      " 'intracellular' 'intracellular ferritin complex'\n",
      " 'intrinsic component of external side of plasma membrane' 'kinetochore'\n",
      " 'lamellar body' 'lamellar body membrane' 'lamellipodium'\n",
      " 'late endosome lumen' 'late endosome membrane' 'lipid droplet'\n",
      " 'lysosomal lumen' 'lysosomal membrane' 'lysosome' 'melanosome lumen'\n",
      " 'melanosome membrane' 'membrane' 'membrane raft' 'microtubule'\n",
      " 'microvillus' 'microvillus membrane' 'midbody' 'mitochondria'\n",
      " 'mitochondrial envelope' 'mitochondrial inner membrane'\n",
      " 'mitochondrial intermembrane space' 'mitochondrial matrix'\n",
      " 'mitochondrial membrane part' 'mitochondrial outer membrane'\n",
      " 'mitochondrion' 'multivesicular body lumen'\n",
      " 'multivesicular body, internal vesicle' 'neuromuscular junction'\n",
      " 'non-motile cilium' 'nuclear body' 'nuclear envelope'\n",
      " 'nuclear inner membrane' 'nuclear lumen' 'nuclear matrix'\n",
      " 'nuclear membrane' 'nuclear speck' 'nucleolus' 'nucleoplasm' 'nucleus'\n",
      " 'outer kinetochore of condensed chromosome'\n",
      " 'outer membrane-bounded periplasmic space' 'pericentriolar material'\n",
      " 'perinuclear region of cytoplasm' 'periplasmic space'\n",
      " 'peroxisomal matrix' 'peroxisomal membrane' 'peroxisome' 'phagocytic cup'\n",
      " 'phagocytic vesicle' 'phagocytic vesicle lumen'\n",
      " 'phagocytic vesicle membrane' 'phagolysosome'\n",
      " 'photoreceptor disc membrane' 'photoreceptor inner segment membrane'\n",
      " 'photoreceptor outer segment membrane' 'plasma membrane'\n",
      " 'plasma membrane (extrinsic to external side of plasma membrane)'\n",
      " 'plasma membrane (extrinsic to internal side of plasma membrane)'\n",
      " 'plasma membrane (integral to plasma membrane)'\n",
      " 'plasma membrane (internal side of plasma membrane)'\n",
      " 'plasma membrane (intrinsic to internal side of plasmamembrane)'\n",
      " 'platelet alpha granule lumen' 'platelet alpha granule membrane'\n",
      " 'platelet dense granule lumen' 'platelet dense granule membrane'\n",
      " 'platelet dense tubular network lumen'\n",
      " 'platelet dense tubular network membrane' 'polysome'\n",
      " 'pre-autophagosomal structure membrane'\n",
      " 'proteinaceous extracellular matrix' 'recycling endosome'\n",
      " 'recycling endosome membrane' 'ribosome' 'sarcoplasmic reticulum'\n",
      " 'sarcoplasmic reticulum lumen' 'sarcoplasmic reticulum membrane'\n",
      " 'secretory granule' 'secretory granule lumen'\n",
      " 'secretory granule membrane' 'signalosome' 'site of double-strand break'\n",
      " 'smooth endoplasmic reticulum' 'specific granule lumen'\n",
      " 'specific granule membrane' 'sperm flagellum' 'sperm plasma membrane'\n",
      " 'spindle' 'spindle microtubule' 'spindle midzone' 'spindle pole'\n",
      " 'synapse' 'synaptic vesicle' 'synaptic vesicle lumen'\n",
      " 'synaptic vesicle membrane' 'tertiary granule lumen'\n",
      " 'tertiary granule membrane' 'tight junction' 'trailing edge'\n",
      " 'trans-Golgi network membrane' 'translation preinitiation complex'\n",
      " 'transport vesicle' 'transport vesicle membrane' 'vesicle'\n",
      " 'viral envelope' 'viral tegument' 'virion membrane']\n"
     ]
    }
   ],
   "source": [
    "def iterate_pathway(p, on_control=None, on_template_reaction=None, on_reaction=None, on_molecule=None):\n",
    "  for r in p['pathwayComponent']:\n",
    "    rr = r\n",
    "    if r['type'] == 'control':\n",
    "      if on_control is not None:\n",
    "        on_control(r)\n",
    "      for m in r['controller']:\n",
    "        if on_molecule is not None:\n",
    "          on_molecule(m, r)\n",
    "    elif r['type'] == 'template_reaction':\n",
    "      if on_template_reaction is not None:\n",
    "        on_template_reaction(r)\n",
    "      for m in r['product']:\n",
    "        if on_molecule is not None:\n",
    "          on_molecule(m, r)\n",
    "    else:\n",
    "      if on_reaction is not None:\n",
    "        on_reaction(r)\n",
    "    if 'left' in rr:\n",
    "      for m in rr['left']:\n",
    "        if on_molecule is not None:\n",
    "          on_molecule(m, r)\n",
    "    if 'right' in rr:\n",
    "      for m in rr['right']:\n",
    "        if on_molecule is not None:\n",
    "          on_molecule(m, r)\n",
    "          \n",
    "\n",
    "cellular_locations = []\n",
    "for p in pathways:\n",
    "  def process_molecule(m, r):\n",
    "    if 'cellularLocation' in m:\n",
    "      cellular_locations.append(m['cellularLocation']) \n",
    "  iterate_pathway(p, on_molecule=process_molecule)\n",
    "print(np.unique(cellular_locations))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "77207\n",
      "38448\n"
     ]
    }
   ],
   "source": [
    "# g = nx.DiGraph()\n",
    "# g = Network(notebook=True, width='100%')\n",
    "\n",
    "n = 0\n",
    "nr = 0\n",
    "un = 0\n",
    "# for r in pathway['pathwayComponent']:\n",
    "#   g.add_node(r['name'])\n",
    "#   rr = r\n",
    "#   if r['type'] == 'catalysis':\n",
    "#     for m in r['controller']:\n",
    "#       if 'gene' in m['entityReference']:\n",
    "#         nr += 1\n",
    "#       n += 1\n",
    "#       g.add_node(m['name'])\n",
    "#       g.add_edge(m['name'], r['name'])\n",
    "#     rr = r['controlled']\n",
    "#   # TODO: conversion direction\n",
    "#   for m in rr['left']:\n",
    "#     if 'gene' in m['entityReference']:\n",
    "#       nr += 1\n",
    "#     n += 1\n",
    "#     g.add_node(m['name'])\n",
    "#     g.add_edge(m['name'], r['name'])\n",
    "#   for m in rr['right']:\n",
    "#     if 'gene' in m['entityReference']:\n",
    "#       nr += 1\n",
    "#     n += 1\n",
    "#     g.add_node(m['name'])\n",
    "#     g.add_edge(r['name'], m['name'])\n",
    "\n",
    "# nx.draw(g)\n",
    "\n",
    "# net = Network(notebook=True, width='100%')\n",
    "# net.from_nx(g)\n",
    "# net.enable_physics(True)\n",
    "# net.show(\"pyvis.html\")\n",
    "\n",
    "for p in filtered_pathways:\n",
    "  def process_molecule(m, r):\n",
    "    global n, nr, un\n",
    "    n += 1\n",
    "    if 'entityReference' not in m:\n",
    "      un += 1\n",
    "      return\n",
    "    if 'gene' in m['entityReference']:\n",
    "      nr += 1\n",
    "  iterate_pathway(p, on_molecule=process_molecule)\n",
    "      \n",
    "print(n)\n",
    "print(nr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import xml.etree.ElementTree as ET\n",
    "from lxml import etree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_gene2uniprot = {}\n",
    "gene_annotations = []\n",
    "with open('../data/uniprot_sprot.xml', 'rb') as f:\n",
    "  context = etree.iterparse(\n",
    "    '../data/uniprot_sprot.xml', \n",
    "    events=(\"end\",), \n",
    "    tag=\"{http://uniprot.org/uniprot}entry\"\n",
    "  )\n",
    "  for action, elem in context:\n",
    "    if action == 'end':\n",
    "#       print(\"%s: %s\" % (action, elem.tag))\n",
    "#       print(elem.xpath(\"./*[name()='dbReference' and @type='GO']/*[name()='property' and @type='term']/@value\"))\n",
    "#       print(elem.xpath(\"./*[name()='comment' and @type='function']/*[name()='text']/text()\"))\n",
    "#       print(etree.tostring(elem, pretty_print=True).decode(\"utf-8\"))\n",
    "#       break\n",
    "      organism = elem.findall('{http://uniprot.org/uniprot}organism')[0] \\\n",
    "            .findall('{http://uniprot.org/uniprot}name')[0].text.strip()\n",
    "      if not 'mus musculus' in organism.lower():\n",
    "        # Memory cleanup\n",
    "        elem.clear()\n",
    "        for ancestor in elem.xpath('ancestor-or-self::*'):\n",
    "            while ancestor.getprevious() is not None:\n",
    "                del ancestor.getparent()[0]\n",
    "        continue\n",
    "      accession = elem.findall('{http://uniprot.org/uniprot}accession')[0].text.strip()\n",
    "      if len(elem.findall('{http://uniprot.org/uniprot}gene')) < 1:\n",
    "        continue\n",
    "      gene_name = elem.findall('{http://uniprot.org/uniprot}gene')[0] \\\n",
    "            .findall('{http://uniprot.org/uniprot}name')[0].text.strip()\n",
    "      mouse_gene2uniprot[gene_name] = accession\n",
    "      \n",
    "      go_terms = elem.xpath(\"./*[name()='dbReference' and @type='GO']/*[name()='property' and @type='term']/@value\")\n",
    "      functions = elem.xpath(\"./*[name()='comment' and @type='function']/*[name()='text']/text()\")\n",
    "      if len(go_terms) > 0 or len(functions) > 0:\n",
    "        gene_annotations.append({\n",
    "          'name': gene_name,\n",
    "          'uniprot_id': accession,\n",
    "          'go_terms': go_terms,\n",
    "          'functions': functions,\n",
    "        })\n",
    "      \n",
    "      # Memory cleanup\n",
    "      elem.clear()\n",
    "      for ancestor in elem.xpath('ancestor-or-self::*'):\n",
    "            while ancestor.getprevious() is not None:\n",
    "                del ancestor.getparent()[0]\n",
    "              \n",
    "#       print('Frog virus 3' in organism)\n",
    "#       print(elem.xpath('.//name', namespaces=context.nsmap))\n",
    "#       print(etree.tostring(elem))\n",
    "#       break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16628\n",
      "16315\n"
     ]
    }
   ],
   "source": [
    "print(len(mouse_gene2uniprot))\n",
    "print(len(gene_annotations))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Lineage</th>\n",
       "      <th>Daphnia pulex</th>\n",
       "      <th>Mus musculus</th>\n",
       "      <th>CS</th>\n",
       "      <th>EL</th>\n",
       "      <th>Trees</th>\n",
       "      <th>phylome</th>\n",
       "      <th>ensembl</th>\n",
       "      <th>orthomcl</th>\n",
       "      <th>treefam</th>\n",
       "      <th>eggnog</th>\n",
       "      <th>hogenom</th>\n",
       "      <th>daphnia</th>\n",
       "      <th>mouse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GO         E9HRC7</td>\n",
       "      <td>GO         Q922V4</td>\n",
       "      <td>1.000</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00/5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E9HRC7</td>\n",
       "      <td>Q922V4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GO         E9GF52</td>\n",
       "      <td>GO         Q3U3J3</td>\n",
       "      <td>0.714</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.60/5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E9GF52</td>\n",
       "      <td>Q3U3J3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GO         E9HAK4</td>\n",
       "      <td>GO         Q3U1N2</td>\n",
       "      <td>1.000</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00/2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E9HAK4</td>\n",
       "      <td>Q3U1N2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GO         E9G2Y1</td>\n",
       "      <td>GO         S4R1P5</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E9G2Y1</td>\n",
       "      <td>S4R1P5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GO         E9FSN2</td>\n",
       "      <td>GO         Q8CFE4</td>\n",
       "      <td>1.000</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00/5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E9FSN2</td>\n",
       "      <td>Q8CFE4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Lineage      Daphnia pulex       Mus musculus     CS  EL  Trees  phylome  \\\n",
       "0      NaN  GO         E9HRC7  GO         Q922V4  1.000   3      7      NaN   \n",
       "1      NaN  GO         E9GF52  GO         Q3U3J3  0.714   3      7      NaN   \n",
       "2      NaN  GO         E9HAK4  GO         Q3U1N2  1.000   3      4      NaN   \n",
       "3      NaN  GO         E9G2Y1  GO         S4R1P5  1.000   1      1      NaN   \n",
       "4      NaN  GO         E9FSN2  GO         Q8CFE4  1.000   3      7      NaN   \n",
       "\n",
       "  ensembl orthomcl  treefam eggnog  hogenom daphnia   mouse  \n",
       "0     NaN   1.00/5      NaN    NaN      NaN  E9HRC7  Q922V4  \n",
       "1     NaN   0.60/5      NaN    NaN      NaN  E9GF52  Q3U3J3  \n",
       "2     NaN   1.00/2      NaN    NaN      NaN  E9HAK4  Q3U1N2  \n",
       "3     NaN      NaN      NaN    NaN      NaN  E9G2Y1  S4R1P5  \n",
       "4     NaN   1.00/5      NaN    NaN      NaN  E9FSN2  Q8CFE4  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_merged = pd.read_csv('./results/merged.csv')\n",
    "daphnia2mouse = pd.read_csv('../data/MetaPhOrs-daphnia2mouse.csv')\n",
    "daphnia2mouse[\"daphnia\"] = daphnia2mouse[\"Daphnia pulex\"].map(lambda g: g.replace('GO', '').strip())\n",
    "daphnia2mouse[\"mouse\"] = daphnia2mouse[\"Mus musculus\"].map(lambda g: g.replace('GO', '').strip())\n",
    "daphnia2mouse.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'name': 'Ywhaq', 'uniprot_id': 'P68254', 'go_terms': ['F:14-3-3 protein binding', 'F:identical protein binding', 'F:ion channel binding', 'F:protein C-terminus binding', 'F:protein domain specific binding', 'F:protein N-terminus binding', 'P:negative regulation of ion transmembrane transport', 'P:negative regulation of transcription, DNA-templated', 'P:protein targeting', 'P:signal transduction', 'P:small GTPase mediated signal transduction']}\n"
     ]
    }
   ],
   "source": [
    "read_genes = { k: '' for k in results_merged['gene'].values }\n",
    "filtered_annotations = list(map(lambda x: \\\n",
    "  {\n",
    "    'name': x['name'],\n",
    "    'uniprot_id': x['uniprot_id'],\n",
    "    'go_terms': list(filter(lambda x: x[:2] in ['F:', 'P:'], x['go_terms'])),\n",
    "  }, gene_annotations))\n",
    "filtered_annotations = list(filter(lambda x: x['name'] in read_genes, filtered_annotations))\n",
    "with open('./results/gene_annotations.json', 'w') as f:\n",
    "  json.dump(filtered_annotations, f)\n",
    "# ga_df = pd.DataFrame(gene_annotations)\n",
    "# ga_df.to_csv('./results/gene_annotations.csv')\n",
    "# ga_df.head()\n",
    "print(filtered_annotations[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in results_merged.iterrows():\n",
    "  if not row['gene'] in mouse_gene2uniprot:\n",
    "    continue\n",
    "  gene_name = mouse_gene2uniprot[row['gene']]\n",
    "  results_merged.loc[index, 'uniprot_mouse'] = gene_name\n",
    "  \n",
    "  homolog_match = daphnia2mouse[daphnia2mouse[\"mouse\"] == gene_name]\n",
    "  if len(homolog_match) < 1:\n",
    "    continue\n",
    "  if homolog_match.iloc[0][\"EL\"] < 2:\n",
    "    # Low number of supporting data sources\n",
    "    continue\n",
    "  homolog = homolog_match.iloc[0][\"daphnia\"]\n",
    "  if homolog.startswith('M!'):\n",
    "    # Not a proper Uniprot id\n",
    "    continue\n",
    "#   print(gene_name)\n",
    "#   print(homolog)\n",
    "  results_merged.loc[index, 'uniprot_daphnia'] = homolog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>index</th>\n",
       "      <th>dataset</th>\n",
       "      <th>start_age</th>\n",
       "      <th>end_age</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue</th>\n",
       "      <th>subtissue</th>\n",
       "      <th>cell_ontology_class</th>\n",
       "      <th>gene</th>\n",
       "      <th>fold_change_log2</th>\n",
       "      <th>expression_at_start_age</th>\n",
       "      <th>expression_at_end_age</th>\n",
       "      <th>expression_max_std_dev</th>\n",
       "      <th>p_value</th>\n",
       "      <th>uniprot_mouse</th>\n",
       "      <th>uniprot_daphnia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>53038</td>\n",
       "      <td>5</td>\n",
       "      <td>droplet</td>\n",
       "      <td>24m</td>\n",
       "      <td>30m</td>\n",
       "      <td>male</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hnf4a</td>\n",
       "      <td>15.215368</td>\n",
       "      <td>4.094030e-09</td>\n",
       "      <td>1.557900e-04</td>\n",
       "      <td>0.000230</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>P49698</td>\n",
       "      <td>E9H7Y7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>11153</td>\n",
       "      <td>3</td>\n",
       "      <td>droplet</td>\n",
       "      <td>3m</td>\n",
       "      <td>24m</td>\n",
       "      <td>male</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hnf4a</td>\n",
       "      <td>-15.065793</td>\n",
       "      <td>1.404470e-04</td>\n",
       "      <td>4.094030e-09</td>\n",
       "      <td>0.000258</td>\n",
       "      <td>7.582662e-73</td>\n",
       "      <td>P49698</td>\n",
       "      <td>E9H7Y7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>17133</td>\n",
       "      <td>15</td>\n",
       "      <td>facs</td>\n",
       "      <td>3m</td>\n",
       "      <td>18m</td>\n",
       "      <td>female</td>\n",
       "      <td>Liver</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Map4k1</td>\n",
       "      <td>15.054170</td>\n",
       "      <td>3.732086e-09</td>\n",
       "      <td>1.270061e-04</td>\n",
       "      <td>0.000393</td>\n",
       "      <td>1.709817e-08</td>\n",
       "      <td>P70218</td>\n",
       "      <td>E9G8P2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>119452</td>\n",
       "      <td>0</td>\n",
       "      <td>facs</td>\n",
       "      <td>3m</td>\n",
       "      <td>24m</td>\n",
       "      <td>male</td>\n",
       "      <td>Brain_Non-Myeloid</td>\n",
       "      <td>Striatum</td>\n",
       "      <td>oligodendrocyte</td>\n",
       "      <td>Ppil2</td>\n",
       "      <td>-14.671002</td>\n",
       "      <td>9.562943e-05</td>\n",
       "      <td>3.664892e-09</td>\n",
       "      <td>0.000178</td>\n",
       "      <td>2.025537e-15</td>\n",
       "      <td>Q9D787</td>\n",
       "      <td>E9H7X4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>119453</td>\n",
       "      <td>1</td>\n",
       "      <td>facs</td>\n",
       "      <td>3m</td>\n",
       "      <td>24m</td>\n",
       "      <td>male</td>\n",
       "      <td>Brain_Non-Myeloid</td>\n",
       "      <td>Striatum</td>\n",
       "      <td>oligodendrocyte</td>\n",
       "      <td>Aldh6a1</td>\n",
       "      <td>-14.523355</td>\n",
       "      <td>8.632671e-05</td>\n",
       "      <td>3.664892e-09</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>1.266913e-15</td>\n",
       "      <td>Q9EQ20</td>\n",
       "      <td>E9GHJ1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0  index  dataset start_age end_age     sex             tissue  \\\n",
       "60       53038      5  droplet       24m     30m    male             Kidney   \n",
       "64       11153      3  droplet        3m     24m    male             Kidney   \n",
       "65       17133     15     facs        3m     18m  female              Liver   \n",
       "82      119452      0     facs        3m     24m    male  Brain_Non-Myeloid   \n",
       "92      119453      1     facs        3m     24m    male  Brain_Non-Myeloid   \n",
       "\n",
       "   subtissue cell_ontology_class     gene  fold_change_log2  \\\n",
       "60       NaN                 NaN    Hnf4a         15.215368   \n",
       "64       NaN                 NaN    Hnf4a        -15.065793   \n",
       "65       NaN                 NaN   Map4k1         15.054170   \n",
       "82  Striatum     oligodendrocyte    Ppil2        -14.671002   \n",
       "92  Striatum     oligodendrocyte  Aldh6a1        -14.523355   \n",
       "\n",
       "    expression_at_start_age  expression_at_end_age  expression_max_std_dev  \\\n",
       "60             4.094030e-09           1.557900e-04                0.000230   \n",
       "64             1.404470e-04           4.094030e-09                0.000258   \n",
       "65             3.732086e-09           1.270061e-04                0.000393   \n",
       "82             9.562943e-05           3.664892e-09                0.000178   \n",
       "92             8.632671e-05           3.664892e-09                0.000160   \n",
       "\n",
       "         p_value uniprot_mouse uniprot_daphnia  \n",
       "60  0.000000e+00        P49698          E9H7Y7  \n",
       "64  7.582662e-73        P49698          E9H7Y7  \n",
       "65  1.709817e-08        P70218          E9G8P2  \n",
       "82  2.025537e-15        Q9D787          E9H7X4  \n",
       "92  1.266913e-15        Q9EQ20          E9GHJ1  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_merged[results_merged['uniprot_daphnia'].notnull()].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_merged.to_csv('./results/merged_augmented.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
