{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gathering pathways, resolving gene homologs and preparing to visualize.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pickle as pkl\n",
    "from importlib import reload\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from owlready2 import *\n",
    "import networkx as nx\n",
    "from pyvis.network import Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "pc = get_ontology(\"../data/PathwayCommons11.reactome.BIOPAX.owl\").load()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load PANTHER gene orthologs\n",
    "```\n",
    "ftp://ftp.pantherdb.org/ortholog/14.1/RefGenomeOrthologs.tar.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>Type</th>\n",
       "      <th>Common ancestor</th>\n",
       "      <th>something else</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARATH|TAIR=AT4G37920|UniProtKB=Q84WN0</td>\n",
       "      <td>ARATH|TAIR=AT1G36320|UniProtKB=Q9C8X8</td>\n",
       "      <td>P</td>\n",
       "      <td>Embryophyta|Magnoliophyta</td>\n",
       "      <td>PTHR31755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HUMAN|HGNC=10663|UniProtKB=O60524</td>\n",
       "      <td>MOUSE|MGI=MGI=1918305|UniProtKB=Q8CCP0</td>\n",
       "      <td>LDO</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>PTHR15239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HUMAN|HGNC=10663|UniProtKB=O60524</td>\n",
       "      <td>RAT|Ensembl=ENSRNOG00000056128|UniProtKB=A0A0G...</td>\n",
       "      <td>LDO</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>PTHR15239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HUMAN|HGNC=10663|UniProtKB=O60524</td>\n",
       "      <td>CHICK|Ensembl=ENSGALG00000012263|UniProtKB=F1N8T0</td>\n",
       "      <td>LDO</td>\n",
       "      <td>Amniota</td>\n",
       "      <td>PTHR15239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HUMAN|HGNC=10663|UniProtKB=O60524</td>\n",
       "      <td>DANRE|Ensembl=ENSDARG00000102859|UniProtKB=A0A...</td>\n",
       "      <td>LDO</td>\n",
       "      <td>Euteleostomi</td>\n",
       "      <td>PTHR15239</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Gene  \\\n",
       "0  ARATH|TAIR=AT4G37920|UniProtKB=Q84WN0   \n",
       "1      HUMAN|HGNC=10663|UniProtKB=O60524   \n",
       "2      HUMAN|HGNC=10663|UniProtKB=O60524   \n",
       "3      HUMAN|HGNC=10663|UniProtKB=O60524   \n",
       "4      HUMAN|HGNC=10663|UniProtKB=O60524   \n",
       "\n",
       "                                            Ortholog Type  \\\n",
       "0              ARATH|TAIR=AT1G36320|UniProtKB=Q9C8X8    P   \n",
       "1             MOUSE|MGI=MGI=1918305|UniProtKB=Q8CCP0  LDO   \n",
       "2  RAT|Ensembl=ENSRNOG00000056128|UniProtKB=A0A0G...  LDO   \n",
       "3  CHICK|Ensembl=ENSGALG00000012263|UniProtKB=F1N8T0  LDO   \n",
       "4  DANRE|Ensembl=ENSDARG00000102859|UniProtKB=A0A...  LDO   \n",
       "\n",
       "             Common ancestor something else  \n",
       "0  Embryophyta|Magnoliophyta      PTHR31755  \n",
       "1           Euarchontoglires      PTHR15239  \n",
       "2           Euarchontoglires      PTHR15239  \n",
       "3                    Amniota      PTHR15239  \n",
       "4               Euteleostomi      PTHR15239  "
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orthologs = pd.read_csv(\n",
    "  '../data/RefGenomeOrthologs', delimiter='\\t', \n",
    "  names=['Gene', 'Ortholog', 'Type', 'Common ancestor', 'something else']\n",
    ")\n",
    "orthologs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build an ortholog -> mouse gene map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_db_name(db):\n",
    "  if db == 'UniProtKB' or db == 'uniprot knowledgebase':\n",
    "    return 'uniprot'\n",
    "  elif db == 'ncbi gene':\n",
    "    return 'ncbi'\n",
    "  return db.lower()\n",
    "\n",
    "\n",
    "def convert_db_id(id):\n",
    "  if ':' in id:\n",
    "    return id.split(':')[-1].lower()\n",
    "  return id.lower()\n",
    "\n",
    "\n",
    "def make_db_ref(db, id):\n",
    "  return '{0}:{1}'.format(convert_db_name(db), convert_db_id(id))\n",
    "\n",
    "\n",
    "try:\n",
    "  # Load precomputed ortholog gene map\n",
    "  with open('../data/gene-map-all2mouse', 'rb') as f:\n",
    "    all2mouse = pkl.load(f)\n",
    "except:\n",
    "  # Compute and store the ortholog gene map\n",
    "  all2mouse = {}\n",
    "  for _, row in orthologs.iterrows():\n",
    "    # Keep only the closest homologs\n",
    "    if row['Type'] != 'LDO':\n",
    "      continue\n",
    "    if not row['Ortholog'].startswith('MOUSE|'):\n",
    "      continue\n",
    "\n",
    "    dst = [\n",
    "      make_db_ref(d.split('=')[0], d.split('=')[-1])\n",
    "      for d in row['Ortholog'].split('|')[1:]\n",
    "    ]\n",
    "\n",
    "    for src in row['Gene'].split('|')[1:]:\n",
    "      src_parts = src.split('=')\n",
    "      db = src_parts[0]\n",
    "      id = src_parts[-1]\n",
    "      all2mouse[make_db_ref(db, id)] = dst\n",
    "\n",
    "  with open('../data/gene-map-all2mouse', 'wb') as f:\n",
    "    pkl.dump(all2mouse, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load mouse gene info from NCBI Gene\n",
    "```\n",
    "ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia/Mus_musculus.gene_info.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_gene_info(path):\n",
    "  \"\"\"Load the NCBI Gene file and returns a {gene_name: info} dictionary\"\"\"\n",
    "  gene_info = pd.read_csv(path, delimiter='\\t')\n",
    "  gene_info_map = {}\n",
    "  for _, row in gene_info.iterrows():\n",
    "    info = {\n",
    "      'name': row['Symbol'],\n",
    "      'type': row['type_of_gene'],\n",
    "      'desc': row['Other_designations'].split('|')[0],\n",
    "    }\n",
    "    ids = [make_db_ref('ncbi', row['GeneID'])]\n",
    "    for xref in row['dbXrefs'].split('|'):\n",
    "      xref_parts = xref.split(':')\n",
    "      db = xref_parts[0]\n",
    "      id = xref_parts[-1]\n",
    "      ids.append(make_db_ref(db, id))\n",
    "    for id in ids:\n",
    "      gene_info_map[id] = info\n",
    "#     for s in row['Synonyms'].split('|'):\n",
    "#       gene_info_map[s] = info\n",
    "  return gene_info, gene_info_map\n",
    "\n",
    "gene_info, gene_info_map = load_gene_info('../data/Mus_musculus.gene_info')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exploring the BIOPAX format here.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{biopax-level3.name, biopax-level3.pathwayComponent, biopax-level3.xref, biopax-level3.displayName, biopax-level3.dataSource, biopax-level3.standardName, biopax-level3.availability}\n",
      "Pyrimidine metabolism\n",
      ".Catalysis_26a2f9367047f173df314be6d019b0cf\n",
      "{biopax-level3.controller, biopax-level3.comment, biopax-level3.displayName, biopax-level3.dataSource, biopax-level3.controlled}\n",
      "rn:R02326_Catalysis\n",
      "\n",
      "[.Protein_807db7d05990fd3bf31ab2a07e82fc60, .Protein_7acd51a37a5a7bc30996301aca90d817]\n",
      "{biopax-level3.name, biopax-level3.comment, biopax-level3.displayName, biopax-level3.dataSource, biopax-level3.standardName, biopax-level3.entityReference}\n",
      "NME\n",
      "uniprot.P56597\n",
      "{biopax-level3.name, biopax-level3.comment, biopax-level3.xref, biopax-level3.organism, biopax-level3.displayName, biopax-level3.standardName}\n",
      "Homo sapiens (human)\n",
      "taxonomy.9606\n",
      "\n",
      ".BiochemicalReaction_6c699bd73dd15be0db5af66fbe386eb0\n",
      "{biopax-level3.conversionDirection, biopax-level3.eCNumber, biopax-level3.participantStoichiometry, biopax-level3.right, biopax-level3.comment, biopax-level3.xref, biopax-level3.displayName, biopax-level3.dataSource, biopax-level3.left}\n",
      "rn:R02326\n",
      "True\n",
      "REVERSIBLE\n",
      "\n",
      "{biopax-level3.name, biopax-level3.comment, biopax-level3.displayName, biopax-level3.dataSource, biopax-level3.standardName, biopax-level3.entityReference}\n",
      "dCDP\n",
      "chebi.CHEBI:28846\n",
      "CHEBI:28846\n",
      "2'-deoxycytidine 5'-(trihydrogen diphosphate)\n",
      "dCDP\n",
      "{biopax-level3.db, biopax-level3.id, biopax-level3.relationshipType}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "Pathway = list(filter(lambda x: x.name == 'Pathway', pc.world.classes()))[0]\n",
    "Catalysis = list(filter(lambda x: x.name == 'Catalysis', pc.world.classes()))[0]\n",
    "BiochemicalReaction = list(filter(lambda x: x.name == 'BiochemicalReaction', pc.world.classes()))[0]\n",
    "for p in Pathway.instances()[100:101]:\n",
    "  print(p.get_properties())\n",
    "  print(p.displayName)\n",
    "  print(p.pathwayComponent[0])\n",
    "  print(p.pathwayComponent[0].get_properties())\n",
    "  print(p.pathwayComponent[0].displayName)\n",
    "  print()\n",
    "  print(p.pathwayComponent[0].controller)\n",
    "  print(p.pathwayComponent[0].controller[0].get_properties())\n",
    "  print(p.pathwayComponent[0].controller[0].displayName)\n",
    "  print(p.pathwayComponent[0].controller[0].entityReference)\n",
    "  print(p.pathwayComponent[0].controller[0].entityReference.get_properties())\n",
    "  print(p.pathwayComponent[0].controller[0].entityReference.organism.displayName)\n",
    "  print(p.pathwayComponent[0].controller[0].entityReference.organism)\n",
    "  print()\n",
    "  print(p.pathwayComponent[0].controlled)\n",
    "  print(p.pathwayComponent[0].controlled.get_properties())\n",
    "  print(p.pathwayComponent[0].controlled.displayName)\n",
    "  print(hasattr(p.pathwayComponent[0].controlled, 'displayName'))\n",
    "  print(p.pathwayComponent[0].controlled.conversionDirection)\n",
    "  print()\n",
    "  print(p.pathwayComponent[0].controlled.right[0].get_properties())\n",
    "  print(p.pathwayComponent[0].controlled.right[0].displayName)\n",
    "  print(p.pathwayComponent[0].controlled.right[0].entityReference)\n",
    "  print(p.pathwayComponent[0].controlled.right[0].entityReference.name)\n",
    "  print(p.pathwayComponent[0].controlled.right[0].entityReference.standardName)\n",
    "  print(p.pathwayComponent[0].controlled.right[0].entityReference.displayName)\n",
    "  print(p.pathwayComponent[0].controlled.right[0].entityReference.xref[0].get_properties())\n",
    "  print()\n",
    "#   print(p.pathwayComponent)\n",
    "\n",
    "#   print(p.xref)\n",
    "#   print(p.xref[0].get_properties())\n",
    "#   print(p.xref[0].db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"hgnc:10663\": [\n",
      "    \"mgi:1918305\",\n",
      "    \"uniprot:q8ccp0\"\n",
      "  ],\n",
      "  \"uniprot:o60524\": [\n",
      "    \"mgi:1918305\",\n",
      "    \"uniprot:q8ccp0\"\n",
      "  ],\n",
      "  \"hgnc:20854\": [\n",
      "    \"mgi:1914066\",\n",
      "    \"uniprot:q9d305\"\n",
      "  ],\n",
      "  \"uniprot:q9h0w7\": [\n",
      "    \"mgi:1914066\",\n",
      "    \"uniprot:q9d305\"\n",
      "  ],\n",
      "  \"hgnc:18049\": [\n",
      "    \"mgi:2152200\",\n",
      "    \"uniprot:q91va0\"\n",
      "  ],\n",
      "  \"uniprot:q08ah1\": [\n",
      "    \"mgi:2152200\",\n",
      "    \"uniprot:q91va0\"\n",
      "  ],\n",
      "  \"hgnc:\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(all2mouse, indent=2)[:400])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse BIOPAX into a more actionable form\n",
    "Extracting pathways and resolving chemical entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_molecule(m):\n",
    "  res = {\n",
    "    'type': 'molecule',\n",
    "    'name': getattr(m, 'displayName', m.name),\n",
    "  }\n",
    "  if hasattr(m, 'entityReference') and m.entityReference is not None:\n",
    "    e = m.entityReference\n",
    "    res['entityReference'] = {\n",
    "      'name': getattr(e, 'displayName', e.name),\n",
    "    }\n",
    "    if hasattr(e, 'xref') and e.xref is not None:\n",
    "      for x in e.xref:\n",
    "        conv_id = '{0}:{1}'.format(convert_db_name(x.db), convert_db_id(x.id))\n",
    "        if conv_id in all2mouse:\n",
    "          for db_ref in all2mouse[conv_id]:\n",
    "            if db_ref in gene_info_map:\n",
    "              res['entityReference']['gene'] = gene_info_map[db_ref]\n",
    "#               print(gene_info_map[db_ref])\n",
    "#           print(all2mouse[conv_id])\n",
    "        # TODO: resolve small molecules\n",
    "        if (x.db == 'ncbi gene') \\\n",
    "          or ( \\\n",
    "            x.db == 'ensembl' \\\n",
    "            and ( \\\n",
    "              'xref' not in res['entityReference'] \\\n",
    "              or res['entityReference']['xref']['db'] != 'ncbi gene' \\\n",
    "            ) \\\n",
    "          ) \\\n",
    "          or ( \\\n",
    "            x.db in ['chebi'] \\\n",
    "            and 'xref' not in res['entityReference'] \\\n",
    "          ):\n",
    "          # NCBI Gene has the top priority and the Ensemlb goes next\n",
    "          res['entityReference']['xref'] = {\n",
    "            'id': x.id,\n",
    "            'db': x.db,\n",
    "          }\n",
    "#       res['entityReference']['xref'] = [\n",
    "#         {\n",
    "#           'id': x.id,\n",
    "#           'db': x.db,\n",
    "# #           'relationshipType': [\n",
    "# #              r.term\n",
    "# #             for r in getattr(x, 'relationshipType', [])\n",
    "# #           ],\n",
    "#         }\n",
    "#         for x in e.xref\n",
    "#       ]\n",
    "#     if hasattr(e, 'organism') and e.organism is not None:\n",
    "#       o = e.organism\n",
    "#       res['entityReference']['organism'] = {\n",
    "#         'name': getattr(o, 'displayName', o.name),\n",
    "#       }\n",
    "#       if 'Homo' not in res['entityReference']['organism']['name']:\n",
    "#         print(res['entityReference']['organism']['name'])\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_reaction(r):\n",
    "  res = {\n",
    "    'type': 'reaction',\n",
    "    'name': getattr(r, 'displayName', r.name),\n",
    "    'comment': getattr(r, 'comment', None),\n",
    "    'conversionDirection': getattr(r, 'conversionDirection', None),\n",
    "  }\n",
    "  if hasattr(r, 'left') and r.left is not None:\n",
    "    res['left'] = [parse_molecule(c) for _, c in enumerate(r.left)]\n",
    "  if hasattr(r, 'right') and r.right is not None:\n",
    "    res['right'] = [parse_molecule(c) for _, c in enumerate(r.left)]\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_catalysis(c):\n",
    "  res = {\n",
    "    'type': 'catalysis',\n",
    "    'name': getattr(c, 'displayName', c.name),\n",
    "    'comment': getattr(c, 'comment', None),\n",
    "  }\n",
    "  if hasattr(c, 'controller') and c.controller is not None:\n",
    "    res['controller'] = [parse_molecule(cc) for cc in c.controller]\n",
    "  if hasattr(c, 'controlled') and c.controlled is not None:\n",
    "    res['controlled'] = parse_reaction(c.controlled)\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_pathway(p):\n",
    "  res = {\n",
    "    'type': 'pathway',\n",
    "    'name': getattr(p, 'displayName', p.name),\n",
    "    'comment': getattr(p, 'comment', None),\n",
    "  }\n",
    "  if hasattr(p, 'pathwayComponent') and p.pathwayComponent is not None:\n",
    "    res['pathwayComponent'] = []\n",
    "    for c in p.pathwayComponent:\n",
    "      if isinstance(c, Catalysis):\n",
    "        res['pathwayComponent'].append(parse_catalysis(c))\n",
    "      else:\n",
    "        res['pathwayComponent'].append(parse_reaction(c))\n",
    "  return res\n",
    "\n",
    "\n",
    "# print(list(pc.world.properties()))\n",
    "\n",
    "for p in Pathway.instances()[100:101]:\n",
    "  pathway = parse_pathway(p)\n",
    "#   print(json.dumps(res, indent=2)[:100000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "843\n",
      "71\n"
     ]
    }
   ],
   "source": [
    "# g = nx.DiGraph()\n",
    "g = Network(notebook=True, width='100%')\n",
    "\n",
    "n = 0\n",
    "nr = 0\n",
    "for r in pathway['pathwayComponent']:\n",
    "  g.add_node(r['name'])\n",
    "  rr = r\n",
    "  if r['type'] == 'catalysis':\n",
    "    for m in r['controller']:\n",
    "      if 'gene' in m['entityReference']:\n",
    "        nr += 1\n",
    "      n += 1\n",
    "      g.add_node(m['name'])\n",
    "      g.add_edge(m['name'], r['name'])\n",
    "    rr = r['controlled']\n",
    "  # TODO: conversion direction\n",
    "  for m in rr['left']:\n",
    "    if 'gene' in m['entityReference']:\n",
    "      nr += 1\n",
    "    n += 1\n",
    "    g.add_node(m['name'])\n",
    "    g.add_edge(m['name'], r['name'])\n",
    "  for m in rr['right']:\n",
    "    if 'gene' in m['entityReference']:\n",
    "      nr += 1\n",
    "    n += 1\n",
    "    g.add_node(m['name'])\n",
    "    g.add_edge(r['name'], m['name'])\n",
    "\n",
    "# nx.draw(g)\n",
    "\n",
    "# net = Network(notebook=True, width='100%')\n",
    "# net.from_nx(g)\n",
    "# net.enable_physics(True)\n",
    "# net.show(\"pyvis.html\")\n",
    "\n",
    "print(n)\n",
    "print(nr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
