{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gathering pathways, resolving gene homologs and preparing to visualize.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy\n",
    "import json\n",
    "import pickle as pkl\n",
    "from importlib import reload\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from owlready2 import *\n",
    "import networkx as nx\n",
    "from pyvis.network import Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pc = get_ontology(\"../data/PathwayCommons11.reactome.BIOPAX.owl\").load()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load PANTHER gene orthologs\n",
    "```\n",
    "ftp://ftp.pantherdb.org/ortholog/14.1/RefGenomeOrthologs.tar.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>Type</th>\n",
       "      <th>Common ancestor</th>\n",
       "      <th>something else</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARATH|TAIR=AT4G37920|UniProtKB=Q84WN0</td>\n",
       "      <td>ARATH|TAIR=AT1G36320|UniProtKB=Q9C8X8</td>\n",
       "      <td>P</td>\n",
       "      <td>Embryophyta|Magnoliophyta</td>\n",
       "      <td>PTHR31755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HUMAN|HGNC=10663|UniProtKB=O60524</td>\n",
       "      <td>MOUSE|MGI=MGI=1918305|UniProtKB=Q8CCP0</td>\n",
       "      <td>LDO</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>PTHR15239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HUMAN|HGNC=10663|UniProtKB=O60524</td>\n",
       "      <td>RAT|Ensembl=ENSRNOG00000056128|UniProtKB=A0A0G...</td>\n",
       "      <td>LDO</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>PTHR15239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HUMAN|HGNC=10663|UniProtKB=O60524</td>\n",
       "      <td>CHICK|Ensembl=ENSGALG00000012263|UniProtKB=F1N8T0</td>\n",
       "      <td>LDO</td>\n",
       "      <td>Amniota</td>\n",
       "      <td>PTHR15239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HUMAN|HGNC=10663|UniProtKB=O60524</td>\n",
       "      <td>DANRE|Ensembl=ENSDARG00000102859|UniProtKB=A0A...</td>\n",
       "      <td>LDO</td>\n",
       "      <td>Euteleostomi</td>\n",
       "      <td>PTHR15239</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Gene  \\\n",
       "0  ARATH|TAIR=AT4G37920|UniProtKB=Q84WN0   \n",
       "1      HUMAN|HGNC=10663|UniProtKB=O60524   \n",
       "2      HUMAN|HGNC=10663|UniProtKB=O60524   \n",
       "3      HUMAN|HGNC=10663|UniProtKB=O60524   \n",
       "4      HUMAN|HGNC=10663|UniProtKB=O60524   \n",
       "\n",
       "                                            Ortholog Type  \\\n",
       "0              ARATH|TAIR=AT1G36320|UniProtKB=Q9C8X8    P   \n",
       "1             MOUSE|MGI=MGI=1918305|UniProtKB=Q8CCP0  LDO   \n",
       "2  RAT|Ensembl=ENSRNOG00000056128|UniProtKB=A0A0G...  LDO   \n",
       "3  CHICK|Ensembl=ENSGALG00000012263|UniProtKB=F1N8T0  LDO   \n",
       "4  DANRE|Ensembl=ENSDARG00000102859|UniProtKB=A0A...  LDO   \n",
       "\n",
       "             Common ancestor something else  \n",
       "0  Embryophyta|Magnoliophyta      PTHR31755  \n",
       "1           Euarchontoglires      PTHR15239  \n",
       "2           Euarchontoglires      PTHR15239  \n",
       "3                    Amniota      PTHR15239  \n",
       "4               Euteleostomi      PTHR15239  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orthologs = pd.read_csv(\n",
    "  '../data/RefGenomeOrthologs', delimiter='\\t', \n",
    "  names=['Gene', 'Ortholog', 'Type', 'Common ancestor', 'something else']\n",
    ")\n",
    "orthologs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build an ortholog -> mouse gene map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_db_name(db):\n",
    "  if db == 'UniProtKB' or db == 'uniprot knowledgebase':\n",
    "    return 'uniprot'\n",
    "  elif db == 'ncbi gene':\n",
    "    return 'ncbi'\n",
    "  elif db == 'hgnc symbol':\n",
    "    return 'hgnc'\n",
    "  return db.lower()\n",
    "\n",
    "\n",
    "def convert_db_id(id):\n",
    "  if ':' in str(id):\n",
    "    return id.split(':')[-1].lower()\n",
    "  return str(id).lower()\n",
    "\n",
    "\n",
    "def make_db_ref(db, id):\n",
    "  return '{0}:{1}'.format(convert_db_name(db), convert_db_id(id))\n",
    "\n",
    "\n",
    "try:\n",
    "  # Load precomputed ortholog gene map\n",
    "  with open('../data/gene-map-all2mouse', 'rb') as f:\n",
    "    all2mouse = pkl.load(f)\n",
    "except:\n",
    "  # Compute and store the ortholog gene map\n",
    "  all2mouse = {}\n",
    "  for _, row in orthologs.iterrows():\n",
    "    # Keep only the closest homologs\n",
    "    if row['Type'] != 'LDO':\n",
    "      continue\n",
    "    if not row['Ortholog'].startswith('MOUSE|'):\n",
    "      continue\n",
    "\n",
    "    dst = [\n",
    "      make_db_ref(d.split('=')[0], d.split('=')[-1])\n",
    "      for d in row['Ortholog'].split('|')[1:]\n",
    "    ]\n",
    "\n",
    "    for src in row['Gene'].split('|')[1:]:\n",
    "      src_parts = src.split('=')\n",
    "      db = src_parts[0]\n",
    "      id = src_parts[-1]\n",
    "      all2mouse[make_db_ref(db, id)] = dst\n",
    "\n",
    "  with open('../data/gene-map-all2mouse', 'wb') as f:\n",
    "    pkl.dump(all2mouse, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load mouse gene info from NCBI Gene\n",
    "```\n",
    "ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia/Mus_musculus.gene_info.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_gene_info(path):\n",
    "  \"\"\"Load the NCBI Gene file and returns a {gene_name: info} dictionary\"\"\"\n",
    "  gene_info = pd.read_csv(path, delimiter='\\t')\n",
    "  gene_info_map = {}\n",
    "  for _, row in gene_info.iterrows():\n",
    "    info = {\n",
    "      'name': row['Symbol'],\n",
    "#       'type': row['type_of_gene'],\n",
    "#       'desc': row['Other_designations'].split('|')[0],\n",
    "    }\n",
    "    ids = [make_db_ref('ncbi', row['GeneID'])]\n",
    "    for xref in row['dbXrefs'].split('|'):\n",
    "      xref_parts = xref.split(':')\n",
    "      db = xref_parts[0]\n",
    "      id = xref_parts[-1]\n",
    "      ids.append(make_db_ref(db, id))\n",
    "    for id in ids:\n",
    "      gene_info_map[id] = info\n",
    "#     for s in row['Synonyms'].split('|'):\n",
    "#       gene_info_map[s] = info\n",
    "  return gene_info, gene_info_map\n",
    "\n",
    "gene_info, gene_info_map = load_gene_info('../data/Mus_musculus.gene_info')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exploring the BIOPAX format here.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['INDIRECT_get_properties', '__class__', '__classcell__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_equivalent_to', '_get_instance_possible_relations', '_get_is_instance_of', '_instance_equivalent_to_changed', '_instance_is_a_changed', '_name', '_set_is_instance_of', 'comment', 'dataSource', 'differents', 'displayName', 'generate_default_name', 'get_equivalent_to', 'get_inverse_properties', 'get_iri', 'get_name', 'get_properties', 'iri', 'is_a', 'is_instance_of', 'name', 'namespace', 'organism', 'pathwayComponent', 'pathwayOrder', 'set_equivalent_to', 'set_iri', 'set_name', 'storid', 'xref']\n",
      "Pathway\n",
      "{biopax-level3.displayName, biopax-level3.pathwayOrder, biopax-level3.organism, biopax-level3.xref, biopax-level3.pathwayComponent, biopax-level3.comment, biopax-level3.dataSource}\n",
      "NTF4 activates NTRK2 (TRKB) signaling\n",
      "reactome.R-HSA-9026510\n",
      "{biopax-level3.displayName, biopax-level3.eCNumber, biopax-level3.xref, biopax-level3.right, biopax-level3.participantStoichiometry, biopax-level3.comment, biopax-level3.dataSource, biopax-level3.left, biopax-level3.conversionDirection}\n",
      "NTF4-bound NTRK2 dimers trans-autophosphorylate\n",
      "\n",
      "{biopax-level3.componentStoichiometry, biopax-level3.displayName, biopax-level3.component, biopax-level3.xref, biopax-level3.cellularLocation, biopax-level3.comment, biopax-level3.dataSource}\n",
      "NTF4:NTRK2 homodimer\n",
      "None\n",
      "--c\n",
      ".Complex_b25a68e543ef2db2d571a00679c5ab9f\n",
      "NTF4:NTRK2\n",
      "--c2\n",
      "{biopax-level3.componentStoichiometry, biopax-level3.displayName, biopax-level3.component, biopax-level3.xref, biopax-level3.name, biopax-level3.cellularLocation, biopax-level3.comment, biopax-level3.dataSource}\n",
      ".Complex_8f7e14ad9cede78207ab6baaa95f0fc9\n",
      "NTF4 homodimer\n",
      "--c3\n",
      "{biopax-level3.displayName, biopax-level3.feature, biopax-level3.name, biopax-level3.xref, biopax-level3.cellularLocation, biopax-level3.comment, biopax-level3.dataSource, biopax-level3.entityReference}\n",
      ".Protein_4c956d35780cd9e1821046ca5ea55a2d\n",
      "NTF4\n",
      "--c2\n",
      "{biopax-level3.displayName, biopax-level3.feature, biopax-level3.name, biopax-level3.xref, biopax-level3.cellularLocation, biopax-level3.comment, biopax-level3.dataSource, biopax-level3.entityReference}\n",
      ".Protein_b5e463297b325abbf5becea6c72b6621\n",
      "TrKB\n",
      "--c\n",
      ".Protein_b5e463297b325abbf5becea6c72b6621\n",
      "TrKB\n",
      "{biopax-level3.idVersion, biopax-level3.id, biopax-level3.db, biopax-level3.comment}\n",
      "R-HSA-9026375\n",
      "reactome\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "Pathway = list(filter(lambda x: x.name == 'Pathway', pc.world.classes()))[0]\n",
    "Interaction = list(filter(lambda x: x.name == 'Interaction', pc.world.classes()))[0]\n",
    "Conversion = list(filter(lambda x: x.name == 'Conversion', pc.world.classes()))[0]\n",
    "Control = list(filter(lambda x: x.name == 'Control', pc.world.classes()))[0]\n",
    "Catalysis = list(filter(lambda x: x.name == 'Catalysis', pc.world.classes()))[0]\n",
    "TemplateReaction = list(filter(lambda x: x.name == 'TemplateReaction', pc.world.classes()))[0]\n",
    "BiochemicalReaction = list(filter(lambda x: x.name == 'BiochemicalReaction', pc.world.classes()))[0]\n",
    "BindingFeature = list(filter(lambda x: x.name == 'BindingFeature', pc.world.classes()))[0]\n",
    "ModificationFeature = list(filter(lambda x: x.name == 'ModificationFeature', pc.world.classes()))[0]\n",
    "Complex = list(filter(lambda x: x.name == 'Complex', pc.world.classes()))[0]\n",
    "for p in Pathway.instances()[0:1]:\n",
    "  print(dir(p))\n",
    "  print(str(p.__class__).replace('biopax-level3.', ''))\n",
    "  print(p.get_properties())\n",
    "  print(p.displayName)\n",
    "  print(p.pathwayComponent[0])\n",
    "  print(p.pathwayComponent[0].get_properties())\n",
    "  print(p.pathwayComponent[0].displayName)\n",
    "  print()\n",
    "  print(p.pathwayComponent[0].left[0].get_properties())\n",
    "  print(p.pathwayComponent[0].left[0].displayName)\n",
    "  print(p.pathwayComponent[0].left[0].entityReference)\n",
    "  for c in p.pathwayComponent[0].left[0].component:\n",
    "    print('--c')\n",
    "    print(c)\n",
    "    print(c.displayName)\n",
    "    if hasattr(c, 'component'):\n",
    "      for c2 in c.component:\n",
    "        print('--c2')\n",
    "        print(c2.get_properties())\n",
    "        print(c2)\n",
    "        print(c2.displayName)\n",
    "        if hasattr(c2, 'component'):\n",
    "          for c3 in c2.component:\n",
    "            print('--c3')\n",
    "            print(c3.get_properties())\n",
    "            print(c3)\n",
    "            print(c3.displayName)\n",
    "  print(p.pathwayComponent[0].left[0].xref[0].get_properties())\n",
    "  for xref in p.pathwayComponent[0].left[0].xref:\n",
    "    print(xref.id)\n",
    "    print(xref.db)\n",
    "#   print(p.pathwayComponent[0].left[0].entityReference.name)\n",
    "#   print(p.pathwayComponent[0].left[0].entityReference.standardName)\n",
    "#   print(p.pathwayComponent[0].left[0].entityReference.displayName)\n",
    "#   print(p.pathwayComponent[0].left[0].entityReference.xref[0].get_properties())\n",
    "\n",
    "#   print(p.pathwayComponent[0].controller)\n",
    "#   print(p.pathwayComponent[0].controller[0].get_properties())\n",
    "#   print(p.pathwayComponent[0].controller[0].displayName)\n",
    "#   print(p.pathwayComponent[0].controller[0].entityReference)\n",
    "#   print(p.pathwayComponent[0].controller[0].entityReference.get_properties())\n",
    "#   print(p.pathwayComponent[0].controller[0].entityReference.organism.displayName)\n",
    "#   print(p.pathwayComponent[0].controller[0].entityReference.organism)\n",
    "#   print()\n",
    "#   print(p.pathwayComponent[0].controlled)\n",
    "#   print(p.pathwayComponent[0].controlled.get_properties())\n",
    "#   print(p.pathwayComponent[0].controlled.displayName)\n",
    "#   print(hasattr(p.pathwayComponent[0].controlled, 'displayName'))\n",
    "#   print(p.pathwayComponent[0].controlled.conversionDirection)\n",
    "  print()\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].get_properties())\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].displayName)\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].entityReference)\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].entityReference.name)\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].entityReference.standardName)\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].entityReference.displayName)\n",
    "#   print(p.pathwayComponent[0].controlled.right[0].entityReference.xref[0].get_properties())\n",
    "  print()\n",
    "#   print(p.pathwayComponent)\n",
    "\n",
    "#   print(p.xref)\n",
    "#   print(p.xref[0].get_properties())\n",
    "#   print(p.xref[0].db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"hgnc:10663\": [\n",
      "    \"mgi:1918305\",\n",
      "    \"uniprot:q8ccp0\"\n",
      "  ],\n",
      "  \"uniprot:o60524\": [\n",
      "    \"mgi:1918305\",\n",
      "    \"uniprot:q8ccp0\"\n",
      "  ],\n",
      "  \"hgnc:20854\": [\n",
      "    \"mgi:1914066\",\n",
      "    \"uniprot:q9d305\"\n",
      "  ],\n",
      "  \"uniprot:q9h0w7\": [\n",
      "    \"mgi:1914066\",\n",
      "    \"uniprot:q9d305\"\n",
      "  ],\n",
      "  \"hgnc:18049\": [\n",
      "    \"mgi:2152200\",\n",
      "    \"uniprot:q91va0\"\n",
      "  ],\n",
      "  \"uniprot:q08ah1\": [\n",
      "    \"mgi:2152200\",\n",
      "    \"uniprot:q91va0\"\n",
      "  ],\n",
      "  \"hgnc:\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(all2mouse, indent=2)[:400])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse BIOPAX into a more actionable form\n",
    "Extracting pathways and resolving chemical entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7085\n",
      "14\n",
      "11980\n",
      "19079\n",
      "19079\n"
     ]
    }
   ],
   "source": [
    "def parse_molecule(m):\n",
    "  res = {\n",
    "    'type': 'molecule',\n",
    "    'name': m.displayName if m.displayName is not None else m.name,\n",
    "#     'name': getattr(m, 'displayName', m.name),\n",
    "#     'name': m.name,\n",
    "  }\n",
    "  if isinstance(m, Complex):\n",
    "    res['type'] = 'complex'\n",
    "    res['component'] = [parse_molecule(cc) for cc in m.component]\n",
    "  if hasattr(m, 'cellularLocation') and m.cellularLocation is not None:\n",
    "    res['cellularLocation'] = m.cellularLocation.term[0]\n",
    "#   if hasattr(m, 'feature') and m.feature is not None:\n",
    "#     res['feature'] = []\n",
    "#     if len(m.feature) > 0:\n",
    "#       for f in m.feature:\n",
    "#         if not isinstance(f, ModificationFeature) and not isinstance(f, BindingFeature):\n",
    "#           continue\n",
    "#         feature = {\n",
    "#           'biopaxType': str(f.__class__).replace('biopax-level3.', ''),\n",
    "#         }\n",
    "#         if hasattr(f, 'modificationType') and f.modificationType is not None:\n",
    "#           feature['modificationType'] = f.modificationType.term[0]\n",
    "#         if hasattr(f, 'featureLocation') and f.featureLocation is not None:\n",
    "#           feature['featureLocation'] = [{\n",
    "#             'sequencePosition': l.sequencePosition,\n",
    "#             'positionStatus': l.positionStatus,\n",
    "#           } for l in f.featureLocation]\n",
    "# #         print(f.get_properties())\n",
    "# #         print(f.modificationType.term[0])\n",
    "# #         print(f.featureLocation[0].get_properties())\n",
    "# #         print(f.featureLocation[0].sequencePosition)\n",
    "# #         print(f.featureLocation[0].positionStatus)\n",
    "# # #         print(f.comment)\n",
    "# #         print(f.featureLocation)\n",
    "# #         print(f.modificationType.term[0])\n",
    "#         res['feature'].append(feature)\n",
    "  if hasattr(m, 'entityReference') and m.entityReference is not None:\n",
    "    e = m.entityReference\n",
    "    res['entityReference'] = {\n",
    "      'name': getattr(e, 'displayName', e.name),\n",
    "    }\n",
    "    if hasattr(e, 'xref') and e.xref is not None:\n",
    "      for x in e.xref:\n",
    "        conv_id = '{0}:{1}'.format(convert_db_name(x.db), convert_db_id(x.id))\n",
    "        if conv_id in all2mouse:\n",
    "          for db_ref in all2mouse[conv_id]:\n",
    "            if db_ref in gene_info_map:\n",
    "              res['entityReference']['gene'] = gene_info_map[db_ref]\n",
    "        # TODO: resolve small molecules\n",
    "        if ('uniprot' in x.db) \\\n",
    "          or ( \\\n",
    "            (x.db == 'ensembl' or x.db == 'ncbi gene') \\\n",
    "            and ( \\\n",
    "              'xref' not in res['entityReference'] \\\n",
    "              or res['entityReference']['xref']['db'] != 'ncbi gene' \\\n",
    "            ) \\\n",
    "          ) \\\n",
    "          or ( \\\n",
    "#             x.db in ['chebi'] \\\n",
    "#             and \n",
    "            'xref' not in res['entityReference'] \\\n",
    "          ):\n",
    "          # NCBI Gene has the top priority and the Ensemlb goes next\n",
    "          res['entityReference']['xref'] = {\n",
    "            'id': convert_db_id(x.id),\n",
    "            'db': convert_db_name(x.db),\n",
    "          }\n",
    "#       res['entityReference']['xref'] = [\n",
    "#         {\n",
    "#           'id': x.id,\n",
    "#           'db': x.db,\n",
    "# #           'relationshipType': [\n",
    "# #              r.term\n",
    "# #             for r in getattr(x, 'relationshipType', [])\n",
    "# #           ],\n",
    "#         }\n",
    "#         for x in e.xref\n",
    "#       ]\n",
    "#     if hasattr(e, 'organism') and e.organism is not None:\n",
    "#       o = e.organism\n",
    "#       res['entityReference']['organism'] = {\n",
    "#         'name': getattr(o, 'displayName', o.name),\n",
    "#       }\n",
    "#       if 'Homo' not in res['entityReference']['organism']['name']:\n",
    "#         print(res['entityReference']['organism']['name'])\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_reaction(r):\n",
    "  res = {\n",
    "    'type': 'reaction',\n",
    "    'name': r.displayName if r.displayName is not None else r.name,\n",
    "#     'name': getattr(r, 'displayName', r.name),\n",
    "#     'comment': getattr(r, 'comment', None),\n",
    "    'conversionDirection': getattr(r, 'conversionDirection', None),\n",
    "  }\n",
    "  if hasattr(r, 'left') and r.left is not None:\n",
    "    res['left'] = [parse_molecule(c) for _, c in enumerate(r.left)]\n",
    "  if hasattr(r, 'right') and r.right is not None:\n",
    "    res['right'] = [parse_molecule(c) for _, c in enumerate(r.right)]\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_control(c):\n",
    "  res = {\n",
    "    'type': 'control',\n",
    "    'name': c.displayName if c.displayName is not None else c.name,\n",
    "#     'name': getattr(c, 'displayName', c.name),\n",
    "#     'comment': getattr(c, 'comment', None),\n",
    "    'controlType': getattr(c, 'controlType', None),\n",
    "  }\n",
    "  if hasattr(c, 'controller') and c.controller is not None:\n",
    "    res['controller'] = [parse_molecule(cc) for cc in c.controller]\n",
    "  if hasattr(c, 'cofactor') and c.cofactor is not None:\n",
    "    res['cofactor'] = [parse_molecule(cc) for cc in c.cofactor]\n",
    "  if hasattr(c, 'controlled') and c.controlled is not None:\n",
    "    res['controlled'] = parse_reaction(c.controlled)\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_template_reaction(c):\n",
    "  res = {\n",
    "    'type': 'template_reaction',\n",
    "    'name': c.displayName if c.displayName is not None else c.name,\n",
    "#     'name': getattr(c, 'displayName', c.name),\n",
    "#     'comment': getattr(c, 'comment', None),\n",
    "    'templateDirection': getattr(c, 'templateDirection', None),\n",
    "  }\n",
    "  if hasattr(c, 'template') and c.template is not None:\n",
    "    res['template'] = parse_molecule(c.template)\n",
    "  if hasattr(c, 'product') and c.product is not None:\n",
    "    res['product'] = [parse_molecule(cc) for cc in c.product]\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_pathway(p):\n",
    "  res = {\n",
    "    'type': 'pathway',\n",
    "    'name': getattr(p, 'displayName', p.name),\n",
    "#     'comment': getattr(p, 'comment', None),\n",
    "  }\n",
    "  if hasattr(p, 'pathwayComponent') and p.pathwayComponent is not None:\n",
    "    res['pathwayComponent'] = []\n",
    "    for c in p.pathwayComponent:\n",
    "      if isinstance(c, Pathway):\n",
    "#         res_list.append(parse_pathway(c))\n",
    "        continue\n",
    "      elif isinstance(c, Control):\n",
    "        res['pathwayComponent'].append(parse_control(c))\n",
    "      elif isinstance(c, TemplateReaction):\n",
    "        res['pathwayComponent'].append(parse_template_reaction(c))\n",
    "      else:\n",
    "        res['pathwayComponent'].append(parse_reaction(c))\n",
    "  return res\n",
    "\n",
    "\n",
    "# print(list(pc.world.properties()))\n",
    "\n",
    "parsed_map = {}\n",
    "pathways = []\n",
    "\n",
    "generic_pathway_template = {\n",
    "  'type': 'pathway',\n",
    "  'name': '<generic>',\n",
    "  'pathwayComponent': [],\n",
    "}\n",
    "\n",
    "def process_reactions(reaction_class, parse_fn):\n",
    "  print(len(reaction_class.instances()))\n",
    "  for c in reaction_class.instances()[:]:\n",
    "    if c.name in parsed_map:\n",
    "      continue\n",
    "    parsed_map[c.name] = ''\n",
    "#     for xref in c.xref:\n",
    "#       print(xref.title)\n",
    "#       print(xref.get_properties())\n",
    "#     print(dir(c))\n",
    "    r = parse_fn(c)\n",
    "  #   print(json.dumps(r, indent=4))\n",
    "    generic_pathway = copy.deepcopy(generic_pathway_template)\n",
    "    generic_pathway['pathwayComponent'].append(r)\n",
    "    pathways.append(generic_pathway) \n",
    "\n",
    "for reaction_class, parse_fn in [\n",
    "  (Control, parse_control),\n",
    "  (TemplateReaction, parse_template_reaction),\n",
    "#   (BiochemicalReaction, parse_reaction),\n",
    "  (Conversion, parse_reaction),\n",
    "  (Interaction, parse_reaction),\n",
    "]:\n",
    "  process_reactions(reaction_class, parse_fn)\n",
    "\n",
    "# for p in Pathway.instances()[:]:\n",
    "#   pathway = parse_pathway(p)\n",
    "# #   print(json.dumps(res, indent=2)[:100000])\n",
    "#   pathways.append(pathway)\n",
    "\n",
    "print(len(pathways))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19079\n"
     ]
    }
   ],
   "source": [
    "# Filter out all the reactions without resolved genes\n",
    "filtered_pathways = []\n",
    "\n",
    "def gene_resolved(m):\n",
    "  if m['type'] == 'complex':\n",
    "    return any([gene_resolved(c) for c in m['component']])\n",
    "  if 'entityReference' not in m:\n",
    "    return False\n",
    "  return 'gene' in m['entityReference']\n",
    "\n",
    "for p in pathways:\n",
    "  filtered_p = copy.deepcopy(p)\n",
    "  filtered_p['pathwayComponent'] = []\n",
    "  for r in p['pathwayComponent']:\n",
    "    resolved = False\n",
    "    rr = r\n",
    "    if r['type'] == 'control':\n",
    "      for m in r['controller']:\n",
    "        resolved = resolved or gene_resolved(m)\n",
    "    elif r['type'] == 'template_reaction':\n",
    "      for m in r['product']:\n",
    "        resolved = resolved or gene_resolved(m)\n",
    "    if 'left' in rr:\n",
    "      for m in rr['left']:\n",
    "        resolved = resolved or gene_resolved(m)\n",
    "    if 'right' in rr:\n",
    "      for m in rr['right']:\n",
    "        resolved = resolved or gene_resolved(m)\n",
    "    if resolved:\n",
    "      filtered_p['pathwayComponent'].append(r)\n",
    "  filtered_pathways.append(filtered_p)\n",
    "\n",
    "print(len(filtered_pathways))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert the parsed data into a graph form"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n",
      "4\n",
      "{\n",
      "    \"nodes\": [\n",
      "        {\n",
      "            \"__id\": 0,\n",
      "            \"type\": \"reaction\",\n",
      "            \"name\": \"TP53 stimulates BNIP3L expression\"\n",
      "        },\n",
      "        {\n",
      "            \"__id\": 1,\n",
      "            \"type\": \"complex\",\n",
      "            \"name\": \"p-S15,S20-TP53 Tetramer\"\n",
      "        },\n",
      "        {\n",
      "            \"__id\": 2,\n",
      "            \"type\": \"molecule\",\n",
      "            \"name\": \"NIX\",\n",
      "            \"cellularLocation\": \"nucleoplasm\",\n",
      "            \"entityReference\": {\n",
      "                \"name\": \"NIX\",\n",
      "                \"gene\": {\n",
      "                    \"name\": \"Bnip3l\"\n",
      "                },\n",
      "                \"xref\": {\n",
      "                    \"id\": \"ensg00000104765\",\n",
      "                    \"db\": \"ensembl\"\n",
      "                }\n",
      "            }\n",
      "        },\n",
      "        {\n",
      "            \"__id\": 3,\n",
      "            \"type\": \"molecule\",\n",
      "            \"name\": \"CBP_HUMAN\",\n",
      "            \"cellularLocation\": \"nucleoplasm\",\n",
      "            \"entityReference\": {\n",
      "                \"name\": \"CBP_HUMAN\",\n",
      "                \"xref\": {\n",
      "                    \"id\": \"q92793\",\n",
      "                    \"db\": \"uniprot\"\n",
      "                },\n",
      "                \"gene\": {\n",
      "                    \"name\": \"Crebbp\"\n",
      "                }\n",
      "            }\n",
      "        },\n",
      "        {\n",
      "            \"__id\": 4,\n",
      "            \"type\": \"molecule\",\n",
      "            \"name\": \"BNI3L_HUMAN\",\n",
      "            \"cellularLocation\": \"mitochondrial outer membrane\",\n",
      "            \"entityReference\": {\n",
      "                \"name\": \"BNI3L_HUMAN\",\n",
      "                \"gene\": {\n",
      "                    \"name\": \"Bnip3l\"\n",
      "                },\n",
      "                \"xref\": {\n",
      "                    \"id\": \"o60238\",\n",
      "                    \"db\": \"uniprot\"\n",
      "                }\n",
      "            }\n",
      "        }\n",
      "    ],\n",
      "    \"edges\": [\n",
      "        [\n",
      "            1,\n",
      "            0,\n",
      "            \"activator\"\n",
      "        ],\n",
      "        [\n",
      "            2,\n",
      "            0,\n",
      "            \"activator\"\n",
      "        ],\n",
      "        [\n",
      "            3,\n",
      "            0,\n",
      "            \"activator\"\n",
      "        ],\n",
      "        [\n",
      "            0,\n",
      "            4,\n",
      "            \"product\"\n",
      "        ]\n",
      "    ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "def location_group(data):\n",
    "  if 'cellularLocation' not in data:\n",
    "    return None\n",
    "  loc = data['cellularLocation']\n",
    "  if any([ l in loc for l in [\n",
    "    'extracellular region',\n",
    "    'external side of plasma membrane',\n",
    "    'cell junction',\n",
    "  ] ]):\n",
    "    return 5\n",
    "  elif any([ l in loc for l in [\n",
    "    'cell outer membrane',\n",
    "    'cell wall',\n",
    "    'cytoplasmic side of plasma membrane',\n",
    "    'plasma membrane',\n",
    "  ] ]):\n",
    "    return 4\n",
    "  elif any([ l in loc for l in [\n",
    "    'mitochondrial',\n",
    "  ] ]):\n",
    "    return 3\n",
    "  elif any([ l in loc for l in [\n",
    "    'cytosol',\n",
    "    'host cell cytosol',\n",
    "  ] ]):\n",
    "    return 0\n",
    "  elif any([ l in loc for l in [\n",
    "    'Golgi',\n",
    "  ] ]):\n",
    "    return -3\n",
    "  elif any([ l in loc for l in [\n",
    "    'endoplasmic reticulum',\n",
    "  ] ]):\n",
    "    return -4\n",
    "  elif any([ l in loc for l in [\n",
    "    'nuclear envelope',\n",
    "    'nucleoplasm',\n",
    "    'chromosome',\n",
    "    'chromosome, centromeric region',\n",
    "  ] ]):\n",
    "    return -5\n",
    "  return None\n",
    "\n",
    "def pathways2graph(pathways):\n",
    "  nodes = []\n",
    "  node_name_map = {}\n",
    "  edges = []\n",
    "  edge_map = {}\n",
    "  \n",
    "  def add_node(data):\n",
    "    name = data['entityReference']['name'] if 'entityReference' in data else data['name']\n",
    "    map_id = f'{name}@{location_group(data)}'\n",
    "    if map_id in node_name_map:\n",
    "      return node_name_map[map_id]\n",
    "    id = len(nodes)\n",
    "    node = {\n",
    "      '__id': id,\n",
    "      'type': data['type'],\n",
    "      'name': name,\n",
    "    }\n",
    "    if data['type'] == 'molecule':\n",
    "      if 'cellularLocation' in data:\n",
    "        node['cellularLocation'] = data['cellularLocation']\n",
    "      if 'entityReference' in data:\n",
    "        node['entityReference'] = data['entityReference']\n",
    "    nodes.append(node)\n",
    "    node_name_map[map_id] = id\n",
    "    return id\n",
    "  \n",
    "  def add_nodes(data):\n",
    "    ids = []\n",
    "    if data['type'] == 'complex':\n",
    "      ids.extend([ add_node(c) for c in data['component'] ])\n",
    "    else:\n",
    "      ids.append(add_node(data))\n",
    "    return list(set(ids))\n",
    "  \n",
    "  def add_edge(source, target, relation):\n",
    "    edge_key = f'{source}|{target}'\n",
    "    if edge_key in edge_map:\n",
    "      return\n",
    "    edges.append([source, target, relation])\n",
    "    edge_map[edge_key] = ''\n",
    "  \n",
    "  control_type2relation = {\n",
    "    'ACTIVATION': 'activator',\n",
    "    'INHIBITION': 'inhibitor',\n",
    "  }\n",
    "  \n",
    "  for pathway in pathways:\n",
    "    for reaction_container in pathway['pathwayComponent']:\n",
    "      reaction = reaction_container\n",
    "      \n",
    "      if reaction_container['type'] == 'control':\n",
    "        reaction = reaction_container['controlled']\n",
    "    \n",
    "      reaction_node_id = add_node(reaction)\n",
    "      if reaction_container['type'] == 'control':\n",
    "        for c in reaction_container['controller']:\n",
    "          molecule_node_ids = add_nodes(c)\n",
    "          for m in molecule_node_ids:\n",
    "            relation = 'controller'\n",
    "            if 'controlType' in reaction_container:\n",
    "              relation = reaction_container['controlType']\n",
    "              for c2r in control_type2relation:\n",
    "                if c2r in relation:\n",
    "                  relation = control_type2relation[c2r]\n",
    "            add_edge(m, reaction_node_id, relation)\n",
    "      elif reaction_container['type'] == 'template_reaction':\n",
    "        if 'template' in reaction_container:\n",
    "          molecule_node_ids = add_nodes(reaction_container['template'])\n",
    "          [ add_edge(m, reaction_node_id, 'template') for m in molecule_node_ids ]\n",
    "        for product in reaction_container['product']:\n",
    "          molecule_node_ids = add_nodes(product)\n",
    "          [ add_edge(reaction_node_id, m, 'product') for m in molecule_node_ids ]\n",
    "          \n",
    "      if 'left' in reaction:\n",
    "        for molecule in reaction['left']:\n",
    "          molecule_node_ids = add_nodes(molecule)\n",
    "          if reaction['conversionDirection'] == 'LEFT_TO_RIGHT':\n",
    "            [ add_edge(m, reaction_node_id, 'reactant') for m in molecule_node_ids ]\n",
    "          elif reaction['conversionDirection'] == 'RIGHT_TO_LEFT':\n",
    "            [ add_edge(reaction_node_id, m, 'product') for m in molecule_node_ids ]\n",
    "          else:\n",
    "            [ add_edge(m, reaction_node_id, 'reactant') for m in molecule_node_ids ]\n",
    "          \n",
    "      if 'right' in reaction:\n",
    "        for molecule in reaction['right']:\n",
    "          molecule_node_ids = add_nodes(molecule)\n",
    "          if reaction['conversionDirection'] == 'LEFT_TO_RIGHT':\n",
    "            [ add_edge(reaction_node_id, m, 'product') for m in molecule_node_ids ]\n",
    "          elif reaction['conversionDirection'] == 'RIGHT_TO_LEFT':\n",
    "            [ add_edge(m, reaction_node_id, 'reactant') for m in molecule_node_ids ]\n",
    "          else:\n",
    "            [ add_edge(reaction_node_id, m, 'product') for m in molecule_node_ids ]\n",
    "        \n",
    "  return {\n",
    "    'nodes': nodes,\n",
    "    'edges': edges,\n",
    "  }\n",
    "\n",
    "print(len(pathways2graph(filtered_pathways[:1])['nodes']))\n",
    "print(len(pathways2graph(filtered_pathways[:1])['edges']))\n",
    "print(json.dumps(pathways2graph(filtered_pathways[:1]), indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19079\n"
     ]
    }
   ],
   "source": [
    "print(len(filtered_pathways))\n",
    "with open('./results/pathways.json', 'w') as f:\n",
    "  json.dump(pathways2graph(filtered_pathways[:]), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"type\": \"pathway\",\n",
      "    \"name\": \"NTF4 activates NTRK2 (TRKB) signaling\",\n",
      "    \"pathwayComponent\": [\n",
      "        {\n",
      "            \"type\": \"reaction\",\n",
      "            \"name\": \"NTF4-bound NTRK2 dimers trans-autophosphorylate\",\n",
      "            \"conversionDirection\": \"LEFT_TO_RIGHT\",\n",
      "            \"left\": [\n",
      "                {\n",
      "                    \"type\": \"complex\",\n",
      "                    \"name\": \"NTF4:NTRK2 homodimer\",\n",
      "                    \"component\": [\n",
      "                        {\n",
      "                            \"type\": \"complex\",\n",
      "                            \"name\": \"NTF4:NTRK2\",\n",
      "                            \"component\": [\n",
      "                                {\n",
      "                                    \"type\": \"complex\",\n",
      "                                    \"name\": \"NTF4 homodimer\",\n",
      "                                    \"component\": [\n",
      "                                        {\n",
      "                                            \"type\": \"molecule\",\n",
      "                                            \"name\": \"NTF4\",\n",
      "                                            \"cellularLocation\": \"extracellular region\",\n",
      "                                            \"entityReference\": {\n",
      "                                                \"name\": \"NTF4_HUMAN\",\n",
      "                                                \"xref\": {\n",
      "                                                    \"id\": \"p34130\",\n",
      "                                                    \"db\": \"uniprot\"\n",
      "                                                },\n",
      "                                                \"gene\": {\n",
      "                                                    \"name\": \"Ntf5\"\n",
      "                                                }\n",
      "                                            }\n",
      "                                        }\n",
      "                                    ],\n",
      "                                    \"cellularLocation\": \"extracellular region\"\n",
      "                                },\n",
      "                                {\n",
      "                                    \"type\": \"molecule\",\n",
      "                                    \"name\": \"TrKB\",\n",
      "                                    \"cellularLocation\": \"plasma membrane\",\n",
      "                                    \"entityReference\": {\n",
      "                                        \"name\": \"NTRK2_HUMAN\",\n",
      "                                        \"xref\": {\n",
      "                                            \"id\": \"q16620\",\n",
      "                                            \"db\": \"uniprot\"\n",
      "                                        },\n",
      "                                        \"gene\": {\n",
      "                                            \"name\": \"Ntrk2\"\n",
      "                                        }\n",
      "                                    }\n",
      "                                }\n",
      "                            ],\n",
      "                            \"cellularLocation\": \"plasma membrane\"\n",
      "                        },\n",
      "                        {\n",
      "                            \"type\": \"molecule\",\n",
      "                            \"name\": \"TrKB\",\n",
      "                            \"cellularLocation\": \"plasma membrane\",\n",
      "                            \"entityReference\": {\n",
      "                                \"name\": \"NTRK2_HUMAN\",\n",
      "                                \"xref\": {\n",
      "                                    \"id\": \"q16620\",\n",
      "                                    \"db\": \"uniprot\"\n",
      "                                },\n",
      "                                \"gene\": {\n",
      "                                    \"name\": \"Ntrk2\"\n",
      "                                }\n",
      "                            }\n",
      "                        }\n",
      "                    ],\n",
      "                    \"cellularLocation\": \"plasma membrane\"\n",
      "                },\n",
      "                {\n",
      "                    \"type\": \"molecule\",\n",
      "                    \"name\": \"ATP\",\n",
      "                    \"cellularLocation\": \"cytosol\",\n",
      "                    \"entityReference\": {\n",
      "                        \"name\": \"ATP\",\n",
      "                        \"xref\": {\n",
      "                            \"id\": \"57299\",\n",
      "                            \"db\": \"chebi\"\n",
      "                        }\n",
      "                    }\n",
      "                }\n",
      "            ],\n",
      "            \"right\": [\n",
      "                {\n",
      "                    \"type\": \"complex\",\n",
      "                    \"name\": \"NT-4:p-5Y-TRKB homodimer\",\n",
      "                    \"component\": [\n",
      "                        {\n",
      "                            \"type\": \"complex\",\n",
      "                            \"name\": \"NT-4:p-5Y-TRKB\",\n",
      "                            \"component\": [\n",
      "                                {\n",
      "                                    \"type\": \"complex\",\n",
      "                                    \"name\": \"NTF4 homodimer\",\n",
      "                                    \"component\": [\n",
      "                                        {\n",
      "                                            \"type\": \"molecule\",\n",
      "                                            \"name\": \"NTF4\",\n",
      "                                            \"cellularLocation\": \"extracellular region\",\n",
      "                                            \"entityReference\": {\n",
      "                                                \"name\": \"NTF4_HUMAN\",\n",
      "                                                \"xref\": {\n",
      "                                                    \"id\": \"p34130\",\n",
      "                                                    \"db\": \"uniprot\"\n",
      "                                                },\n",
      "                                                \"gene\": {\n",
      "                                                    \"name\": \"Ntf5\"\n",
      "                                                }\n",
      "                                            }\n",
      "                                        }\n",
      "                                    ],\n",
      "                                    \"cellularLocation\": \"extracellular region\"\n",
      "                                },\n",
      "                                {\n",
      "                                    \"type\": \"molecule\",\n",
      "                                    \"name\": \"p-5Y-TrKB\",\n",
      "                                    \"cellularLocation\": \"plasma membrane\",\n",
      "                                    \"entityReference\": {\n",
      "                                        \"name\": \"NTRK2_HUMAN\",\n",
      "                                        \"xref\": {\n",
      "                                            \"id\": \"q16620\",\n",
      "                                            \"db\": \"uniprot\"\n",
      "                                        },\n",
      "                                        \"gene\": {\n",
      "                                            \"name\": \"Ntrk2\"\n",
      "                                        }\n",
      "                                    }\n",
      "                                }\n",
      "                            ],\n",
      "                            \"cellularLocation\": \"plasma membrane\"\n",
      "                        },\n",
      "                        {\n",
      "                            \"type\": \"molecule\",\n",
      "                            \"name\": \"p-5Y-TrKB\",\n",
      "                            \"cellularLocation\": \"plasma membrane\",\n",
      "                            \"entityReference\": {\n",
      "                                \"name\": \"NTRK2_HUMAN\",\n",
      "                                \"xref\": {\n",
      "                                    \"id\": \"q16620\",\n",
      "                                    \"db\": \"uniprot\"\n",
      "                                },\n",
      "                                \"gene\": {\n",
      "                                    \"name\": \"Ntrk2\"\n",
      "                                }\n",
      "                            }\n",
      "                        }\n",
      "                    ],\n",
      "                    \"cellularLocation\": \"plasma membrane\"\n",
      "                },\n",
      "                {\n",
      "                    \"type\": \"molecule\",\n",
      "                    \"name\": \"ADP\",\n",
      "                    \"cellularLocation\": \"cytosol\",\n",
      "                    \"entityReference\": {\n",
      "                        \"name\": \"ADP\",\n",
      "                        \"xref\": {\n",
      "                            \"id\": \"87518\",\n",
      "                            \"db\": \"chebi\"\n",
      "                        }\n",
      "                    }\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"type\": \"reaction\",\n",
      "            \"name\": \"NTF4-bound NTRK2 dimerizes\",\n",
      "            \"conversionDirection\": \"LEFT_TO_RIGHT\",\n",
      "            \"left\": [\n",
      "                {\n",
      "                    \"type\": \"complex\",\n",
      "                    \"name\": \"NTF4:NTRK2\",\n",
      "                    \"component\": [\n",
      "                        {\n",
      "                            \"type\": \"complex\",\n",
      "                            \"name\": \"NTF4 homodimer\",\n",
      "                            \"component\": [\n",
      "                                {\n",
      "                                    \"type\": \"molecule\",\n",
      "                                    \"name\": \"NTF4\",\n",
      "                                    \"cellularLocation\": \"extracellular region\",\n",
      "                                    \"entityReference\": {\n",
      "                                        \"name\": \"NTF4_HUMAN\",\n",
      "                                        \"xref\": {\n",
      "                                            \"id\": \"p34130\",\n",
      "                                            \"db\": \"uniprot\"\n",
      "                                        },\n",
      "                                        \"gene\": {\n",
      "                                            \"name\": \"Ntf5\"\n",
      "                                        }\n",
      "                                    }\n",
      "                                }\n",
      "                            ],\n",
      "                            \"cellularLocation\": \"extracellular region\"\n",
      "                        },\n",
      "                        {\n",
      "                            \"type\": \"molecule\",\n",
      "                            \"name\": \"TrKB\",\n",
      "                            \"cellularLocation\": \"plasma membrane\",\n",
      "                            \"entityReference\": {\n",
      "                                \"name\": \"NTRK2_HUMAN\",\n",
      "                                \"xref\": {\n",
      "                                    \"id\": \"q16620\",\n",
      "                                    \"db\": \"uniprot\"\n",
      "                                },\n",
      "                                \"gene\": {\n",
      "                                    \"name\": \"Ntrk2\"\n",
      "                                }\n",
      "                            }\n",
      "                        }\n",
      "                    ],\n",
      "                    \"cellularLocation\": \"plasma membrane\"\n",
      "                },\n",
      "                {\n",
      "                    \"type\": \"molecule\",\n",
      "                    \"name\": \"TrKB\",\n",
      "                    \"cellularLocation\": \"plasma membrane\",\n",
      "                    \"entityReference\": {\n",
      "                        \"name\": \"NTRK2_HUMAN\",\n",
      "                        \"xref\": {\n",
      "                            \"id\": \"q16620\",\n",
      "                            \"db\": \"uniprot\"\n",
      "                        },\n",
      "                        \"gene\": {\n",
      "                            \"name\": \"Ntrk2\"\n",
      "                        }\n",
      "                    }\n",
      "                }\n",
      "            ],\n",
      "            \"right\": [\n",
      "                {\n",
      "                    \"type\": \"complex\",\n",
      "                    \"name\": \"NTF4:NTRK2 homodimer\",\n",
      "                    \"component\": [\n",
      "                        {\n",
      "                            \"type\": \"complex\",\n",
      "                            \"name\": \"NTF4:NTRK2\",\n",
      "                            \"component\": [\n",
      "                                {\n",
      "                                    \"type\": \"complex\",\n",
      "                                    \"name\": \"NTF4 homodimer\",\n",
      "                                    \"component\": [\n",
      "                                        {\n",
      "                                            \"type\": \"molecule\",\n",
      "                                            \"name\": \"NTF4\",\n",
      "                                            \"cellularLocation\": \"extracellular region\",\n",
      "                                            \"entityReference\": {\n",
      "                                                \"name\": \"NTF4_HUMAN\",\n",
      "                                                \"xref\": {\n",
      "                                                    \"id\": \"p34130\",\n",
      "                                                    \"db\": \"uniprot\"\n",
      "                                                },\n",
      "                                                \"gene\": {\n",
      "                                                    \"name\": \"Ntf5\"\n",
      "                                                }\n",
      "                                            }\n",
      "                                        }\n",
      "                                    ],\n",
      "                                    \"cellularLocation\": \"extracellular region\"\n",
      "                                },\n",
      "                                {\n",
      "                                    \"type\": \"molecule\",\n",
      "                                    \"name\": \"TrKB\",\n",
      "                                    \"cellularLocation\": \"plasma membrane\",\n",
      "                                    \"entityReference\": {\n",
      "                                        \"name\": \"NTRK2_HUMAN\",\n",
      "                                        \"xref\": {\n",
      "                                            \"id\": \"q16620\",\n",
      "                                            \"db\": \"uniprot\"\n",
      "                                        },\n",
      "                                        \"gene\": {\n",
      "                                            \"name\": \"Ntrk2\"\n",
      "                                        }\n",
      "                                    }\n",
      "                                }\n",
      "                            ],\n",
      "                            \"cellularLocation\": \"plasma membrane\"\n",
      "                        },\n",
      "                        {\n",
      "                            \"type\": \"molecule\",\n",
      "                            \"name\": \"TrKB\",\n",
      "                            \"cellularLocation\": \"plasma membrane\",\n",
      "                            \"entityReference\": {\n",
      "                                \"name\": \"NTRK2_HUMAN\",\n",
      "                                \"xref\": {\n",
      "                                    \"id\": \"q16620\",\n",
      "                                    \"db\": \"uniprot\"\n",
      "                                },\n",
      "                                \"gene\": {\n",
      "                                    \"name\": \"Ntrk2\"\n",
      "                                }\n",
      "                            }\n",
      "                        }\n",
      "                    ],\n",
      "                    \"cellularLocation\": \"plasma membrane\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"type\": \"reaction\",\n",
      "            \"name\": \"NTF4 binds NTRK2\",\n",
      "            \"conversionDirection\": \"LEFT_TO_RIGHT\",\n",
      "            \"left\": [\n",
      "                {\n",
      "                    \"type\": \"complex\",\n",
      "                    \"name\": \"NTF4 homodimer\",\n",
      "                    \"component\": [\n",
      "                        {\n",
      "                            \"type\": \"molecule\",\n",
      "                            \"name\": \"NTF4\",\n",
      "                            \"cellularLocation\": \"extracellular region\",\n",
      "                            \"entityReference\": {\n",
      "                                \"name\": \"NTF4_HUMAN\",\n",
      "                                \"xref\": {\n",
      "                                    \"id\": \"p34130\",\n",
      "                                    \"db\": \"uniprot\"\n",
      "                                },\n",
      "                                \"gene\": {\n",
      "                                    \"name\": \"Ntf5\"\n",
      "                                }\n",
      "                            }\n",
      "                        }\n",
      "                    ],\n",
      "                    \"cellularLocation\": \"extracellular region\"\n",
      "                },\n",
      "                {\n",
      "                    \"type\": \"molecule\",\n",
      "                    \"name\": \"TrKB\",\n",
      "                    \"cellularLocation\": \"plasma membrane\",\n",
      "                    \"entityReference\": {\n",
      "                        \"name\": \"NTRK2_HUMAN\",\n",
      "                        \"xref\": {\n",
      "                            \"id\": \"q16620\",\n",
      "                            \"db\": \"uniprot\"\n",
      "                        },\n",
      "                        \"gene\": {\n",
      "                            \"name\": \"Ntrk2\"\n",
      "                        }\n",
      "                    }\n",
      "                }\n",
      "            ],\n",
      "            \"right\": [\n",
      "                {\n",
      "                    \"type\": \"complex\",\n",
      "                    \"name\": \"NTF4:NTRK2\",\n",
      "                    \"component\": [\n",
      "                        {\n",
      "                            \"type\": \"complex\",\n",
      "                            \"name\": \"NTF4 homodimer\",\n",
      "                            \"component\": [\n",
      "                                {\n",
      "                                    \"type\": \"molecule\",\n",
      "                                    \"name\": \"NTF4\",\n",
      "                                    \"cellularLocation\": \"extracellular region\",\n",
      "                                    \"entityReference\": {\n",
      "                                        \"name\": \"NTF4_HUMAN\",\n",
      "                                        \"xref\": {\n",
      "                                            \"id\": \"p34130\",\n",
      "                                            \"db\": \"uniprot\"\n",
      "                                        },\n",
      "                                        \"gene\": {\n",
      "                                            \"name\": \"Ntf5\"\n",
      "                                        }\n",
      "                                    }\n",
      "                                }\n",
      "                            ],\n",
      "                            \"cellularLocation\": \"extracellular region\"\n",
      "                        },\n",
      "                        {\n",
      "                            \"type\": \"molecule\",\n",
      "                            \"name\": \"TrKB\",\n",
      "                            \"cellularLocation\": \"plasma membrane\",\n",
      "                            \"entityReference\": {\n",
      "                                \"name\": \"NTRK2_HUMAN\",\n",
      "                                \"xref\": {\n",
      "                                    \"id\": \"q16620\",\n",
      "                                    \"db\": \"uniprot\"\n",
      "                                },\n",
      "                                \"gene\": {\n",
      "                                    \"name\": \"Ntrk2\"\n",
      "                                }\n",
      "                            }\n",
      "                        }\n",
      "                    ],\n",
      "                    \"cellularLocation\": \"plasma membrane\"\n",
      "                }\n",
      "            ]\n",
      "        }\n",
      "    ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# print(len(pathways2graph(filtered_pathways[:])['nodes']))\n",
    "# print(len(pathways2graph(filtered_pathways[:])['edges']))\n",
    "print(json.dumps(filtered_pathways[0], indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['COPII-coated ER to Golgi transport vesicle'\n",
      " 'ER to Golgi transport vesicle membrane' 'Golgi lumen' 'Golgi membrane'\n",
      " 'Golgi-associated vesicle lumen' 'Golgi-associated vesicle membrane'\n",
      " 'autophagosome' 'autophagosome membrane' 'axonemal microtubule'\n",
      " 'azurophil granule lumen' 'azurophil granule membrane' 'cell junction'\n",
      " 'cell outer membrane' 'cell wall' 'chromosome'\n",
      " 'chromosome, centromeric region' 'ciliary basal body' 'ciliary base'\n",
      " 'ciliary membrane' 'ciliary tip' 'cilium'\n",
      " 'clathrin-coated endocytic vesicle'\n",
      " 'clathrin-coated endocytic vesicle membrane'\n",
      " 'clathrin-coated vesicle membrane'\n",
      " 'clathrin-sculpted acetylcholine transport vesicle lumen'\n",
      " 'clathrin-sculpted gamma-aminobutyric acid transport vesicle lumen'\n",
      " 'clathrin-sculpted gamma-aminobutyric acid transport vesicle membrane'\n",
      " 'clathrin-sculpted monoamine transport vesicle lumen'\n",
      " 'clathrin-sculpted monoamine transport vesicle membrane'\n",
      " 'cornified envelope' 'cytoplasm' 'cytoplasmic side of plasma membrane'\n",
      " 'cytoplasmic vesicle lumen' 'cytoplasmic vesicle membrane' 'cytosol'\n",
      " 'early endosome' 'early endosome lumen' 'early endosome membrane'\n",
      " 'endocytic vesicle lumen' 'endocytic vesicle membrane'\n",
      " 'endolysosome membrane' 'endoplasmic reticulum lumen'\n",
      " 'endoplasmic reticulum membrane'\n",
      " 'endoplasmic reticulum quality control compartment'\n",
      " 'endoplasmic reticulum-Golgi intermediate compartment'\n",
      " 'endoplasmic reticulum-Golgi intermediate compartment membrane'\n",
      " 'endosome' 'endosome lumen' 'endosome membrane'\n",
      " 'external side of plasma membrane' 'extracellular exosome'\n",
      " 'extracellular region' 'ficolin-1-rich granule lumen'\n",
      " 'ficolin-1-rich granule membrane' 'host cell cytosol'\n",
      " 'integral component of cytoplasmic side of endoplasmic reticulum membrane'\n",
      " 'integral component of lumenal side of endoplasmic reticulum membrane'\n",
      " 'lamellar body' 'lamellar body membrane' 'late endosome lumen'\n",
      " 'late endosome membrane' 'lipid droplet' 'lysosomal lumen'\n",
      " 'lysosomal membrane' 'melanosome lumen' 'melanosome membrane'\n",
      " 'microvillus' 'microvillus membrane' 'mitochondrial inner membrane'\n",
      " 'mitochondrial intermembrane space' 'mitochondrial matrix'\n",
      " 'mitochondrial outer membrane' 'multivesicular body lumen'\n",
      " 'multivesicular body, internal vesicle' 'nuclear envelope' 'nucleoplasm'\n",
      " 'periplasmic space' 'peroxisomal matrix' 'peroxisomal membrane'\n",
      " 'phagocytic cup' 'phagocytic vesicle' 'phagocytic vesicle lumen'\n",
      " 'phagocytic vesicle membrane' 'phagolysosome'\n",
      " 'photoreceptor disc membrane' 'photoreceptor outer segment membrane'\n",
      " 'plasma membrane' 'platelet alpha granule lumen'\n",
      " 'platelet alpha granule membrane' 'platelet dense granule lumen'\n",
      " 'platelet dense granule membrane' 'platelet dense tubular network lumen'\n",
      " 'platelet dense tubular network membrane'\n",
      " 'pre-autophagosomal structure membrane' 'recycling endosome'\n",
      " 'recycling endosome membrane' 'sarcoplasmic reticulum lumen'\n",
      " 'sarcoplasmic reticulum membrane' 'secretory granule lumen'\n",
      " 'secretory granule membrane' 'smooth endoplasmic reticulum'\n",
      " 'specific granule lumen' 'specific granule membrane' 'sperm flagellum'\n",
      " 'sperm plasma membrane' 'synaptic vesicle' 'synaptic vesicle membrane'\n",
      " 'tertiary granule lumen' 'tertiary granule membrane'\n",
      " 'trans-Golgi network membrane' 'transport vesicle'\n",
      " 'transport vesicle membrane' 'viral envelope' 'viral tegument'\n",
      " 'virion membrane']\n"
     ]
    }
   ],
   "source": [
    "def iterate_pathway(p, on_control=None, on_template_reaction=None, on_reaction=None, on_molecule=None):\n",
    "  for r in p['pathwayComponent']:\n",
    "    rr = r\n",
    "    if r['type'] == 'control':\n",
    "      if on_control is not None:\n",
    "        on_control(r)\n",
    "      for m in r['controller']:\n",
    "        if on_molecule is not None:\n",
    "          on_molecule(m, r)\n",
    "    elif r['type'] == 'template_reaction':\n",
    "      if on_template_reaction is not None:\n",
    "        on_template_reaction(r)\n",
    "      for m in r['product']:\n",
    "        if on_molecule is not None:\n",
    "          on_molecule(m, r)\n",
    "    else:\n",
    "      if on_reaction is not None:\n",
    "        on_reaction(r)\n",
    "    if 'left' in rr:\n",
    "      for m in rr['left']:\n",
    "        if on_molecule is not None:\n",
    "          on_molecule(m, r)\n",
    "    if 'right' in rr:\n",
    "      for m in rr['right']:\n",
    "        if on_molecule is not None:\n",
    "          on_molecule(m, r)\n",
    "          \n",
    "\n",
    "cellular_locations = []\n",
    "for p in pathways:\n",
    "  def process_molecule(m, r):\n",
    "    if 'cellularLocation' in m:\n",
    "      cellular_locations.append(m['cellularLocation']) \n",
    "  iterate_pathway(p, on_molecule=process_molecule)\n",
    "print(np.unique(cellular_locations))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17055\n",
      "8174\n"
     ]
    }
   ],
   "source": [
    "# g = nx.DiGraph()\n",
    "# g = Network(notebook=True, width='100%')\n",
    "\n",
    "n = 0\n",
    "nr = 0\n",
    "un = 0\n",
    "# for r in pathway['pathwayComponent']:\n",
    "#   g.add_node(r['name'])\n",
    "#   rr = r\n",
    "#   if r['type'] == 'catalysis':\n",
    "#     for m in r['controller']:\n",
    "#       if 'gene' in m['entityReference']:\n",
    "#         nr += 1\n",
    "#       n += 1\n",
    "#       g.add_node(m['name'])\n",
    "#       g.add_edge(m['name'], r['name'])\n",
    "#     rr = r['controlled']\n",
    "#   # TODO: conversion direction\n",
    "#   for m in rr['left']:\n",
    "#     if 'gene' in m['entityReference']:\n",
    "#       nr += 1\n",
    "#     n += 1\n",
    "#     g.add_node(m['name'])\n",
    "#     g.add_edge(m['name'], r['name'])\n",
    "#   for m in rr['right']:\n",
    "#     if 'gene' in m['entityReference']:\n",
    "#       nr += 1\n",
    "#     n += 1\n",
    "#     g.add_node(m['name'])\n",
    "#     g.add_edge(r['name'], m['name'])\n",
    "\n",
    "# nx.draw(g)\n",
    "\n",
    "# net = Network(notebook=True, width='100%')\n",
    "# net.from_nx(g)\n",
    "# net.enable_physics(True)\n",
    "# net.show(\"pyvis.html\")\n",
    "\n",
    "for p in filtered_pathways:\n",
    "  def process_molecule(m, r):\n",
    "    global n, nr, un\n",
    "    n += 1\n",
    "    if 'entityReference' not in m:\n",
    "      un += 1\n",
    "      return\n",
    "    if 'gene' in m['entityReference']:\n",
    "      nr += 1\n",
    "  iterate_pathway(p, on_molecule=process_molecule)\n",
    "      \n",
    "print(n)\n",
    "print(nr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import xml.etree.ElementTree as ET\n",
    "from lxml import etree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_gene2uniprot = {}\n",
    "with open('../data/uniprot_sprot.xml', 'rb') as f:\n",
    "  context = etree.iterparse(\n",
    "    '../data/uniprot_sprot.xml', \n",
    "    events=(\"end\",), \n",
    "    tag=\"{http://uniprot.org/uniprot}entry\"\n",
    "  )\n",
    "  for action, elem in context:\n",
    "    if action == 'end':\n",
    "#       print(\"%s: %s\" % (action, elem.tag))\n",
    "      organism = elem.findall('{http://uniprot.org/uniprot}organism')[0] \\\n",
    "            .findall('{http://uniprot.org/uniprot}name')[0].text.strip()\n",
    "      if not 'mus musculus' in organism.lower():\n",
    "        # Memory cleanup\n",
    "        elem.clear()\n",
    "        for ancestor in elem.xpath('ancestor-or-self::*'):\n",
    "            while ancestor.getprevious() is not None:\n",
    "                del ancestor.getparent()[0]\n",
    "        continue\n",
    "      accession = elem.findall('{http://uniprot.org/uniprot}accession')[0].text.strip()\n",
    "      if len(elem.findall('{http://uniprot.org/uniprot}gene')) < 1:\n",
    "        continue\n",
    "      gene_name = elem.findall('{http://uniprot.org/uniprot}gene')[0] \\\n",
    "            .findall('{http://uniprot.org/uniprot}name')[0].text.strip()\n",
    "      mouse_gene2uniprot[gene_name] = accession\n",
    "      \n",
    "      # Memory cleanup\n",
    "      elem.clear()\n",
    "      for ancestor in elem.xpath('ancestor-or-self::*'):\n",
    "            while ancestor.getprevious() is not None:\n",
    "                del ancestor.getparent()[0]\n",
    "              \n",
    "#       print('Frog virus 3' in organism)\n",
    "#       print(elem.xpath('.//name', namespaces=context.nsmap))\n",
    "#       print(etree.tostring(elem))\n",
    "#       break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16628"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(mouse_gene2uniprot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Lineage</th>\n",
       "      <th>Daphnia pulex</th>\n",
       "      <th>Mus musculus</th>\n",
       "      <th>CS</th>\n",
       "      <th>EL</th>\n",
       "      <th>Trees</th>\n",
       "      <th>phylome</th>\n",
       "      <th>ensembl</th>\n",
       "      <th>orthomcl</th>\n",
       "      <th>treefam</th>\n",
       "      <th>eggnog</th>\n",
       "      <th>hogenom</th>\n",
       "      <th>daphnia</th>\n",
       "      <th>mouse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GO         E9HRC7</td>\n",
       "      <td>GO         Q922V4</td>\n",
       "      <td>1.000</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00/5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E9HRC7</td>\n",
       "      <td>Q922V4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GO         E9GF52</td>\n",
       "      <td>GO         Q3U3J3</td>\n",
       "      <td>0.714</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.60/5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E9GF52</td>\n",
       "      <td>Q3U3J3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GO         E9HAK4</td>\n",
       "      <td>GO         Q3U1N2</td>\n",
       "      <td>1.000</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00/2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E9HAK4</td>\n",
       "      <td>Q3U1N2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GO         E9G2Y1</td>\n",
       "      <td>GO         S4R1P5</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E9G2Y1</td>\n",
       "      <td>S4R1P5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GO         E9FSN2</td>\n",
       "      <td>GO         Q8CFE4</td>\n",
       "      <td>1.000</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00/5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E9FSN2</td>\n",
       "      <td>Q8CFE4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Lineage      Daphnia pulex       Mus musculus     CS  EL  Trees  phylome  \\\n",
       "0      NaN  GO         E9HRC7  GO         Q922V4  1.000   3      7      NaN   \n",
       "1      NaN  GO         E9GF52  GO         Q3U3J3  0.714   3      7      NaN   \n",
       "2      NaN  GO         E9HAK4  GO         Q3U1N2  1.000   3      4      NaN   \n",
       "3      NaN  GO         E9G2Y1  GO         S4R1P5  1.000   1      1      NaN   \n",
       "4      NaN  GO         E9FSN2  GO         Q8CFE4  1.000   3      7      NaN   \n",
       "\n",
       "  ensembl orthomcl  treefam eggnog  hogenom daphnia   mouse  \n",
       "0     NaN   1.00/5      NaN    NaN      NaN  E9HRC7  Q922V4  \n",
       "1     NaN   0.60/5      NaN    NaN      NaN  E9GF52  Q3U3J3  \n",
       "2     NaN   1.00/2      NaN    NaN      NaN  E9HAK4  Q3U1N2  \n",
       "3     NaN      NaN      NaN    NaN      NaN  E9G2Y1  S4R1P5  \n",
       "4     NaN   1.00/5      NaN    NaN      NaN  E9FSN2  Q8CFE4  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_merged = pd.read_csv('./results/merged.csv')\n",
    "daphnia2mouse = pd.read_csv('../data/MetaPhOrs-daphnia2mouse.csv')\n",
    "daphnia2mouse[\"daphnia\"] = daphnia2mouse[\"Daphnia pulex\"].map(lambda g: g.replace('GO', '').strip())\n",
    "daphnia2mouse[\"mouse\"] = daphnia2mouse[\"Mus musculus\"].map(lambda g: g.replace('GO', '').strip())\n",
    "daphnia2mouse.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in results_merged.iterrows():\n",
    "  if not row['gene'] in mouse_gene2uniprot:\n",
    "    continue\n",
    "  gene_name = mouse_gene2uniprot[row['gene']]\n",
    "  results_merged.loc[index, 'uniprot_mouse'] = gene_name\n",
    "  \n",
    "  homolog_match = daphnia2mouse[daphnia2mouse[\"mouse\"] == gene_name]\n",
    "  if len(homolog_match) < 1:\n",
    "    continue\n",
    "  if homolog_match.iloc[0][\"EL\"] < 2:\n",
    "    # Low number of supporting data sources\n",
    "    continue\n",
    "  homolog = homolog_match.iloc[0][\"daphnia\"]\n",
    "  if homolog.startswith('M!'):\n",
    "    # Not a proper Uniprot id\n",
    "    continue\n",
    "#   print(gene_name)\n",
    "#   print(homolog)\n",
    "  results_merged.loc[index, 'uniprot_daphnia'] = homolog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>index</th>\n",
       "      <th>dataset</th>\n",
       "      <th>start_age</th>\n",
       "      <th>end_age</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue</th>\n",
       "      <th>subtissue</th>\n",
       "      <th>cell_ontology_class</th>\n",
       "      <th>gene</th>\n",
       "      <th>fold_change_log2</th>\n",
       "      <th>expression_at_start_age</th>\n",
       "      <th>expression_at_end_age</th>\n",
       "      <th>expression_max_std_dev</th>\n",
       "      <th>p_value</th>\n",
       "      <th>uniprot_mouse</th>\n",
       "      <th>uniprot_daphnia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>19245</td>\n",
       "      <td>5</td>\n",
       "      <td>droplet</td>\n",
       "      <td>24m</td>\n",
       "      <td>30m</td>\n",
       "      <td>male</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hnf4a</td>\n",
       "      <td>15.215368</td>\n",
       "      <td>4.094030e-09</td>\n",
       "      <td>1.557900e-04</td>\n",
       "      <td>0.000230</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>P49698</td>\n",
       "      <td>E9H7Y7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>2428</td>\n",
       "      <td>3</td>\n",
       "      <td>droplet</td>\n",
       "      <td>3m</td>\n",
       "      <td>24m</td>\n",
       "      <td>male</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hnf4a</td>\n",
       "      <td>-15.065793</td>\n",
       "      <td>1.404470e-04</td>\n",
       "      <td>4.094030e-09</td>\n",
       "      <td>0.000258</td>\n",
       "      <td>7.582662e-73</td>\n",
       "      <td>P49698</td>\n",
       "      <td>E9H7Y7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>5398</td>\n",
       "      <td>15</td>\n",
       "      <td>facs</td>\n",
       "      <td>3m</td>\n",
       "      <td>18m</td>\n",
       "      <td>female</td>\n",
       "      <td>Liver</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Map4k1</td>\n",
       "      <td>15.054170</td>\n",
       "      <td>3.732086e-09</td>\n",
       "      <td>1.270061e-04</td>\n",
       "      <td>0.000393</td>\n",
       "      <td>1.709817e-08</td>\n",
       "      <td>P70218</td>\n",
       "      <td>E9G8P2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>42689</td>\n",
       "      <td>0</td>\n",
       "      <td>facs</td>\n",
       "      <td>3m</td>\n",
       "      <td>24m</td>\n",
       "      <td>male</td>\n",
       "      <td>Brain_Non-Myeloid</td>\n",
       "      <td>Striatum</td>\n",
       "      <td>oligodendrocyte</td>\n",
       "      <td>Ppil2</td>\n",
       "      <td>-14.671002</td>\n",
       "      <td>9.562943e-05</td>\n",
       "      <td>3.664892e-09</td>\n",
       "      <td>0.000178</td>\n",
       "      <td>2.025537e-15</td>\n",
       "      <td>Q9D787</td>\n",
       "      <td>E9H7X4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>42690</td>\n",
       "      <td>1</td>\n",
       "      <td>facs</td>\n",
       "      <td>3m</td>\n",
       "      <td>24m</td>\n",
       "      <td>male</td>\n",
       "      <td>Brain_Non-Myeloid</td>\n",
       "      <td>Striatum</td>\n",
       "      <td>oligodendrocyte</td>\n",
       "      <td>Aldh6a1</td>\n",
       "      <td>-14.523355</td>\n",
       "      <td>8.632671e-05</td>\n",
       "      <td>3.664892e-09</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>1.266913e-15</td>\n",
       "      <td>Q9EQ20</td>\n",
       "      <td>E9GHJ1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0  index  dataset start_age end_age     sex             tissue  \\\n",
       "60       19245      5  droplet       24m     30m    male             Kidney   \n",
       "64        2428      3  droplet        3m     24m    male             Kidney   \n",
       "65        5398     15     facs        3m     18m  female              Liver   \n",
       "82       42689      0     facs        3m     24m    male  Brain_Non-Myeloid   \n",
       "92       42690      1     facs        3m     24m    male  Brain_Non-Myeloid   \n",
       "\n",
       "   subtissue cell_ontology_class     gene  fold_change_log2  \\\n",
       "60       NaN                 NaN    Hnf4a         15.215368   \n",
       "64       NaN                 NaN    Hnf4a        -15.065793   \n",
       "65       NaN                 NaN   Map4k1         15.054170   \n",
       "82  Striatum     oligodendrocyte    Ppil2        -14.671002   \n",
       "92  Striatum     oligodendrocyte  Aldh6a1        -14.523355   \n",
       "\n",
       "    expression_at_start_age  expression_at_end_age  expression_max_std_dev  \\\n",
       "60             4.094030e-09           1.557900e-04                0.000230   \n",
       "64             1.404470e-04           4.094030e-09                0.000258   \n",
       "65             3.732086e-09           1.270061e-04                0.000393   \n",
       "82             9.562943e-05           3.664892e-09                0.000178   \n",
       "92             8.632671e-05           3.664892e-09                0.000160   \n",
       "\n",
       "         p_value uniprot_mouse uniprot_daphnia  \n",
       "60  0.000000e+00        P49698          E9H7Y7  \n",
       "64  7.582662e-73        P49698          E9H7Y7  \n",
       "65  1.709817e-08        P70218          E9G8P2  \n",
       "82  2.025537e-15        Q9D787          E9H7X4  \n",
       "92  1.266913e-15        Q9EQ20          E9GHJ1  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_merged[results_merged['uniprot_daphnia'].notnull()].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_merged.to_csv('./results/merged_augmented.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
